---
title: "Copy number analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{04: Copy number analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r}
library(maftools)
```

Note: This vignette not evaluated.

## Introduction

`maftools` provides set of functions to facilitate using [ASCAT](https://github.com/VanLoo-lab/ascat) for tumor-normal WGS pairs.
Although there exists [ascatNgs](https://github.com/cancerit/ascatNgs), it requires installation of perl and C modules. `maftools` bypasses these requirements entirely within R with the C code baked in. However, `maftools` only generates the required count inputs and the downstream analyses has to be done with [ASCAT](https://github.com/VanLoo-lab/ascat). 

If you find the below packages useful, please cite the ASCAT publication

------------------------------------------------------------------------
***Van Loo P, Nordgard SH, Lingj√¶rde OC, et al. Allele-specific copy number analysis of tumors. Proc Natl Acad Sci U S A. 2010;107(39):16910-16915. doi:10.1073/pnas.1009843107***
------------------------------------------------------------------------

### Step-1: Get nucleotide counts for genetic markers with `gtMarkers`

Below command will generate two tsv files `tumor_nucleotide_counts.tsv` and `normal_nucleotide_counts.tsv` that can be used for downstream analysis. Note that the function will process ~900K SNPs from [Affymetrix Genome-Wide Human SNP 6.0 Array](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL6801). The process can be sped up by increasing `nthreads` which will launch each chromosome on a separate thread. 
Currently `hg19` and `hg38` are supported.

```{r, eval=FALSE}
#Matched normal BAM files are strongly recommended
counts = maftools::gtMarkers(t_bam = "tumor.bam", n_bam = "normal.bam", build = "hg19")
```

### Step-2: Prepare input files for ASCAT with `prepAscat()`

#### Tumor-Normal pair

Below command will filter SNPs with low coverage (default <30), estimate BAF, logR, and generates the input files for ASCAT.

In addition, if [ASCAT](https://github.com/VanLoo-lab/ascat) is installed, it will run `ASCAT::ascat.loadData()` and `ASCAT::ascat.plotRawData()` for you and returns the ASCAT object that can be further processed with further downstream ASCAT functions.

```{r, eval=FALSE}
library(ASCAT)
ascat.bc = maftools::prepAscat(t_counts = "tumor_nucleotide_counts.tsv", n_counts = "normal_nucleotide_counts.tsv", sample_name = "tumor")

# Markers: 901235
# Removed 3072 duplicated loci
# Markers > 30: 25246
# ------
# Counts file: normal_nucleotide_counts.tsv
# Markers: 901235
# Removed 3072 duplicated loci
# Markers > 30: 31387
# ------
# Final number SNPs: 23765
# Generated following files:
# tumor.tumour.BAF.txt
# tumor.tumour.logR.txt
# tumor.normal.BAF.txt
# tumor.normal.logR.txt
# ------
# Running ASCAT::ascat.loadData:
# [1] Reading Tumor LogR data...
# [1] Reading Tumor BAF data...
# [1] Reading Germline LogR data...
# [1] Reading Germline BAF data...
# [1] Registering SNP locations...
# [1] Splitting genome in distinct chunks...
# Running ASCAT::ascat.plotRawData:
# [1] Plotting tumor data
# [1] Plotting germline data
# Returned ASCAT object
```

The returned `ASCAT` object can be passed to downstream ASCAT functions:

```{r, eval=FALSE}
ascat.bc = ASCAT::ascat.aspcf(ascat.bc)
ASCAT::ascat.plotSegmentedData(ascat.bc)
ascat.output = ASCAT::ascat.runAscat(ascat.bc) 
```

#### Tumor only

```{r, eval=FALSE}
ascat.bc = maftools::prepAscat_t(t_counts = "tumor_nucleotide_counts.tsv", sample_name = "tumoronly")

# Library sizes:
# Tumor: 1239964831
# Counts file: tumor_nucleotide_counts.tsv
# Markers: 930104
# Removed 15 duplicated loci
# Markers > 30: 829579
# ------
# Median depth of coverage: 59
# Generated following files:
# tumoronly.tumour.BAF.txt
# tumoronly.tumour.logR.txt
# ------
# Running ASCAT::ascat.loadData:
# [1] Reading Tumor LogR data...
# [1] Reading Tumor BAF data...
# [1] Registering SNP locations...
# [1] Splitting genome in distinct chunks...
# Running ASCAT::ascat.plotRawData()
# [1] Plotting tumor data
# Returned ASCAT object!
```

The returned `ASCAT` object can be processed with _ASCAT without matched normal data protocol_:

```{r, eval=FALSE}
ascat.gg = ASCAT::ascat.predictGermlineGenotypes(ascat.bc) 
ascat.bc = ASCAT::ascat.aspcf(ascat.bc,ascat.gg=ascat.gg) 
ASCAT::ascat.plotSegmentedData(ascat.bc)
ascat.output = ASCAT::ascat.runAscat(ascat.bc) 
```

### CBS segmentation

Alternatively, tumor logR files generated by `prep_ascat()` can be processed with `segment_logR()` function that performs circular binary segmentation using [DNAcopy](https://bioconductor.org/packages/release/bioc/html/DNAcopy.html) and plots the results

```{r, eval=FALSE}
maftools::segmentLogR(tumor_logR = "tumor.tumour.logR.txt", sample_name = "tumor")

# Analyzing: tumor 
#   current chromosome: 1 
#   current chromosome: 2 
#   current chromosome: 3 
#   current chromosome: 4 
#   current chromosome: 5 
#   current chromosome: 6 
#   current chromosome: 7 
#   current chromosome: 8 
#   current chromosome: 9 
#   current chromosome: 10 
#   current chromosome: 11 
#   current chromosome: 12 
#   current chromosome: 13 
#   current chromosome: 14 
#   current chromosome: 15 
#   current chromosome: 16 
#   current chromosome: 17 
#   current chromosome: 18 
#   current chromosome: 19 
#   current chromosome: 20 
#   current chromosome: 21 
#   current chromosome: 22 
#   current chromosome: MT 
#   current chromosome: X 
#   current chromosome: Y 
# Segments are written to: tumor_cbs.seg
```
