<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>maftools : Summarize, Analyze and Visualize MAF Files • maftools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="maftools : Summarize, Analyze and Visualize MAF Files">
<meta property="og:description" content="maftools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maftools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.4.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/maftools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/oncoplots.html">Customizing oncoplots</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PoisonAlien/maftools/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="maftools_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>maftools : Summarize, Analyze and Visualize MAF Files</h1>
                        <h4 class="author">Anand Mayakonda</h4>
            
            <h4 class="date">2020-08-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/PoisonAlien/maftools/blob/master/vignettes/maftools.Rmd"><code>vignettes/maftools.Rmd</code></a></small>
      <div class="hidden name"><code>maftools.Rmd</code></div>

    </div>

    
    
<p><img src="maftools.png" height="100" width="250" style="position:absolute;top:30px;right:0px;"></p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>With advances in Cancer Genomics, <a href="https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/">Mutation Annotation Format</a> (MAF) is being widely accepted and used to store somatic variants detected. <a href="http://cancergenome.nih.gov">The Cancer Genome Atlas</a> Project has sequenced over 30 different cancers with sample size of each cancer type being over 200. <a href="https://wiki.nci.nih.gov/display/TCGA/TCGA+MAF+Files">Resulting data</a> consisting of somatic variants are stored in the form of <a href="https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/">Mutation Annotation Format</a>. This package attempts to summarize, analyze, annotate and visualize MAF files in an efficient manner from either TCGA sources or any in-house studies as long as the data is in MAF format.</p>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>If you find this tool useful, please cite:</p>
<hr>
<p><strong><em>Mayakonda A, Lin DC, Assenov Y, Plass C, Koeffler HP. 2018. Maftools: efficient and comprehensive analysis of somatic variants in cancer. <a href="https://doi.org/10.1101/gr.239244.118">Genome Resarch</a> <a href="https://www.ncbi.nlm.nih.gov/pubmed/30341162">PMID: 30341162</a></em></strong></p>
<hr>
</div>
</div>
<div id="generating-maf-files" class="section level1">
<h1 class="hasAnchor">
<a href="#generating-maf-files" class="anchor"></a>Generating MAF files</h1>
<ul>
<li><p>For VCF files or simple tabular files, easy option is to use <a href="https://github.com/mskcc/vcf2maf">vcf2maf</a> utility which will annotate VCFs, prioritize transcripts, and generates an MAF. Recent updates to gatk has also enabled <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035889931-Funcotator-Information-and-Tutorial">funcotator</a> to genrate MAF files.</p></li>
<li><p>If you’re using <a href="http://annovar.openbioinformatics.org/en/latest/">ANNOVAR</a> for variant annotations, maftools has a handy function <code>annovarToMaf</code> for converting tabular annovar outputs to MAF.</p></li>
</ul>
</div>
<div id="maf-field-requirements" class="section level1">
<h1 class="hasAnchor">
<a href="#maf-field-requirements" class="anchor"></a>MAF field requirements</h1>
<p>MAF files contain many fields ranging from chromosome names to cosmic annotations. However most of the analysis in maftools uses following fields.</p>
<ul>
<li><p>Mandatory fields: <strong>Hugo_Symbol, Chromosome, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, Variant_Classification, Variant_Type and Tumor_Sample_Barcode</strong>.</p></li>
<li><p>Recommended optional fields: non MAF specific fields containing VAF (Variant Allele Frequecy) and amino acid change information.</p></li>
</ul>
<p>Complete specification of MAF files can be found on <a href="https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/">NCI GDC documentation page</a>.</p>
<p>This vignette demonstrates the usage and application of maftools on an example MAF file from TCGA LAML cohort <a href="#references">1</a>.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="kw">if</span> (!<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="st">"BiocManager"</span>))
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"maftools"</span>)</pre></body></html></div>
</div>
<div id="overview-of-the-package" class="section level1">
<h1 class="hasAnchor">
<a href="#overview-of-the-package" class="anchor"></a>Overview of the package</h1>
<p>maftools functions can be categorized into mainly Visualization and Analysis modules. Each of these functions and a short description is summarized as shown below. Usage is simple, just read your MAF file with <code>read.maf</code> (along with copy-number data if available) and pass the resulting MAF object to the desired function for plotting or analysis.</p>
<p><img src="overview.png"></p>
</div>
<div id="reading-and-summarizing-maf-files" class="section level1">
<h1 class="hasAnchor">
<a href="#reading-and-summarizing-maf-files" class="anchor"></a>Reading and summarizing maf files</h1>
<div id="required-input-files" class="section level2">
<h2 class="hasAnchor">
<a href="#required-input-files" class="anchor"></a>Required input files</h2>
<ul>
<li>an MAF file - can be gz compressed. Required.</li>
<li>an optional but recommended clinical data associated with each sample/Tumor_Sample_Barcode in MAF.</li>
<li>an optional copy number data if available. Can be GISTIC output or a custom table containing sample names, gene names and copy-number status (<code>Amp</code> or <code>Del</code>).</li>
</ul>
</div>
<div id="reading-maf-files-" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-maf-files-" class="anchor"></a>Reading MAF files.</h2>
<p><code>read.maf</code> function reads MAF files, summarizes it in various ways and stores it as an MAF object. Even though MAF file is alone enough, it is recommended to provide annotations associated with samples in MAF. One can also integrate copy number data if available.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">maftools</span>)</pre></body></html></div>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="co">#path to TCGA LAML MAF file</span>
<span class="no">laml.maf</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'tcga_laml.maf.gz'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'maftools'</span>)
<span class="co">#clinical information containing survival information and histology. This is optional</span>
<span class="no">laml.clin</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'tcga_laml_annot.tsv'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'maftools'</span>)

<span class="no">laml</span> <span class="kw">=</span> <span class="fu"><a href="../reference/read.maf.html">read.maf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml.maf</span>, <span class="kw">clinicalData</span> <span class="kw">=</span> <span class="no">laml.clin</span>)</pre></body></html></div>
<pre><code>## -Reading
## -Validating
## -Silent variants: 475 
## -Summarizing
## -Processing clinical data
## -Finished in 0.327s elapsed (0.288s cpu)</code></pre>
</div>
<div id="maf-object" class="section level2">
<h2 class="hasAnchor">
<a href="#maf-object" class="anchor"></a>MAF object</h2>
<p>Summarized MAF file is stored as an MAF object. MAF object contains main maf file, summarized data and any associated sample annotations.</p>
<p>There are accessor methods to access the useful slots from MAF object.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co">#Typing laml shows basic summary of MAF file.</span>
<span class="no">laml</span></pre></body></html></div>
<pre><code>## An object of class  MAF 
##                    ID          summary  Mean Median
##  1:        NCBI_Build               37    NA     NA
##  2:            Center genome.wustl.edu    NA     NA
##  3:           Samples              193    NA     NA
##  4:            nGenes             1241    NA     NA
##  5:   Frame_Shift_Del               52 0.269      0
##  6:   Frame_Shift_Ins               91 0.472      0
##  7:      In_Frame_Del               10 0.052      0
##  8:      In_Frame_Ins               42 0.218      0
##  9: Missense_Mutation             1342 6.953      7
## 10: Nonsense_Mutation              103 0.534      0
## 11:       Splice_Site               92 0.477      0
## 12:             total             1732 8.974      9</code></pre>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="co">#Shows sample summry.</span>
<span class="fu"><a href="../reference/getSampleSummary.html">getSampleSummary</a></span>(<span class="no">laml</span>)
<span class="co">#Shows gene summary.</span>
<span class="fu"><a href="../reference/getGeneSummary.html">getGeneSummary</a></span>(<span class="no">laml</span>)
<span class="co">#shows clinical data associated with samples</span>
<span class="fu"><a href="../reference/getClinicalData.html">getClinicalData</a></span>(<span class="no">laml</span>)
<span class="co">#Shows all fields in MAF</span>
<span class="fu"><a href="../reference/getFields.html">getFields</a></span>(<span class="no">laml</span>)
<span class="co">#Writes maf summary to an output file with basename laml.</span>
<span class="fu"><a href="../reference/write.mafSummary.html">write.mafSummary</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">basename</span> <span class="kw">=</span> <span class="st">'laml'</span>)</pre></body></html></div>
</div>
</div>
<div id="visualization" class="section level1">
<h1 class="hasAnchor">
<a href="#visualization" class="anchor"></a>Visualization</h1>
<div id="plotting-maf-summary-" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-maf-summary-" class="anchor"></a>Plotting MAF summary.</h2>
<p>We can use <code>plotmafSummary</code> to plot the summary of the maf file, which displays number of variants in each sample as a stacked barplot and variant types as a boxplot summarized by Variant_Classification.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="../reference/plotmafSummary.html">plotmafSummary</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">rmOutlier</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">addStat</span> <span class="kw">=</span> <span class="st">'median'</span>, <span class="kw">dashboard</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">titvRaw</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
</div>
<div id="oncoplots" class="section level2">
<h2 class="hasAnchor">
<a href="#oncoplots" class="anchor"></a>Oncoplots</h2>
<div id="drawing-oncoplots" class="section level3">
<h3 class="hasAnchor">
<a href="#drawing-oncoplots" class="anchor"></a>Drawing oncoplots</h3>
<p>Better representation of maf file can be shown as oncoplots, also known as waterfall plots.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="co">#oncoplot for top ten mutated genes.</span>
<span class="fu"><a href="../reference/oncoplot.html">oncoplot</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">top</span> <span class="kw">=</span> <span class="fl">10</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-7-1.png" width="960" style="display: block; margin: auto auto auto 0;"></p>
<p>NOTE: Variants annotated as <code>Multi_Hit</code> are those genes which are mutated more than once in the same sample.</p>
<p>For more details on customisation see the <a href="http://bioconductor.org/packages/devel/bioc/vignettes/maftools/inst/doc/oncoplots.html">Customizing oncoplots</a> vignette.</p>
</div>
</div>
<div id="transition-and-transversions-" class="section level2">
<h2 class="hasAnchor">
<a href="#transition-and-transversions-" class="anchor"></a>Transition and Transversions.</h2>
<p><code>titv</code> function classifies SNPs into <a href="http://www.mun.ca/biology/scarr/Transitions_vs_Transversions.html">Transitions and Transversions</a> and returns a list of summarized tables in various ways. Summarized data can also be visualized as a boxplot showing overall distribution of six different conversions and as a stacked barplot showing fraction of conversions in each sample.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">laml.titv</span> <span class="kw">=</span> <span class="fu"><a href="../reference/titv.html">titv</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">plot</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">useSyn</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co">#plot titv summary</span>
<span class="fu"><a href="../reference/plotTiTv.html">plotTiTv</a></span>(<span class="kw">res</span> <span class="kw">=</span> <span class="no">laml.titv</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="lollipop-plots-for-amino-acid-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#lollipop-plots-for-amino-acid-changes" class="anchor"></a>Lollipop plots for amino acid changes</h2>
<p><code>lollipopPlot</code> function requires us to have amino acid changes information in the maf file. However MAF files have no clear guidelines on naming the field for amino acid changes, with different studies having different field (or column) names for amino acid changes. By default, <code>lollipopPlot</code> looks for column <code>AAChange</code>, and if its not found in the MAF file, it prints all available fields with a warning message. For below example, MAF file contains amino acid changes under a field/column name ‘Protein_Change’. We will manually specify this using argument <code>AACol</code>.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co">#lollipop plot for DNMT3A, which is one of the most frequent mutated gene in Leukemia.</span>
<span class="fu"><a href="../reference/lollipopPlot.html">lollipopPlot</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">gene</span> <span class="kw">=</span> <span class="st">'DNMT3A'</span>, <span class="kw">AACol</span> <span class="kw">=</span> <span class="st">'Protein_Change'</span>, <span class="kw">showMutationRate</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>## 3 transcripts available. Use arguments refSeqID or proteinID to manually specify tx name.</code></pre>
<pre><code>##      HGNC refseq.ID protein.ID aa.length
## 1: DNMT3A NM_175629  NP_783328       912
## 2: DNMT3A NM_022552  NP_072046       912
## 3: DNMT3A NM_153759  NP_715640       723</code></pre>
<pre><code>## Using longer transcript NM_175629 for now.</code></pre>
<pre><code>## Removed 3 mutations for which AA position was not available</code></pre>
<p><img src="maftools_files/figure-html/unnamed-chunk-9-1.png" width="576" style="display: block; margin: auto auto auto 0;"></p>
<p>By default lollipopPlot uses the longest isoform of the gene.</p>
<div id="labelling-points-" class="section level3">
<h3 class="hasAnchor">
<a href="#labelling-points-" class="anchor"></a>Labelling points.</h3>
<p>We can also label points on the <code>lollipopPlot</code> using argument <code>labelPos</code>. If <code>labelPos</code> is set to ‘all’, all of the points are highlighted.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu"><a href="../reference/lollipopPlot.html">lollipopPlot</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">gene</span> <span class="kw">=</span> <span class="st">'DNMT3A'</span>, <span class="kw">showDomainLabel</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">labelPos</span> <span class="kw">=</span> <span class="fl">882</span>)</pre></body></html></div>
<pre><code>## Assuming protein change information are stored under column Protein_Change. Use argument AACol to override if necessary.</code></pre>
<pre><code>## 3 transcripts available. Use arguments refSeqID or proteinID to manually specify tx name.</code></pre>
<pre><code>##      HGNC refseq.ID protein.ID aa.length
## 1: DNMT3A NM_175629  NP_783328       912
## 2: DNMT3A NM_022552  NP_072046       912
## 3: DNMT3A NM_153759  NP_715640       723</code></pre>
<pre><code>## Using longer transcript NM_175629 for now.</code></pre>
<pre><code>## Removed 3 mutations for which AA position was not available</code></pre>
<p><img src="maftools_files/figure-html/unnamed-chunk-10-1.png" width="576" style="display: block; margin: auto auto auto 0;"></p>
</div>
</div>
<div id="rainfall-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#rainfall-plots" class="anchor"></a>Rainfall plots</h2>
<p>Cancer genomes, especially solid tumors are characterized by genomic loci with localized hyper-mutations <a href="#references">5</a>. Such hyper mutated genomic regions can be visualized by plotting inter variant distance on a linear genomic scale. These plots generally called rainfall plots and we can draw such plots using <code>rainfallPlot</code>. If <code>detectChangePoints</code> is set to TRUE, <code>rainfall</code> plot also highlights regions where potential changes in inter-event distances are located.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">brca</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"brca.maf.gz"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="no">brca</span> <span class="kw">=</span> <span class="fu"><a href="../reference/read.maf.html">read.maf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">brca</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu"><a href="../reference/rainfallPlot.html">rainfallPlot</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">brca</span>, <span class="kw">detectChangePoints</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">pointSize</span> <span class="kw">=</span> <span class="fl">0.6</span>)</pre></body></html></div>
<pre><code>## Processing TCGA-A8-A08B..</code></pre>
<pre><code>## Kataegis detected at:</code></pre>
<pre><code>##    Chromosome Start_Position End_Position nMuts Avg_intermutation_dist Size
## 1:          8       98129348     98133560     7               702.0000 4212
## 2:          8       98398549     98403536     9               623.3750 4987
## 3:          8       98453076     98456466     9               423.7500 3390
## 4:          8      124090377    124096810    22               306.3333 6433
## 5:         12       97436055     97439705     7               608.3333 3650
## 6:         17       29332072     29336153     8               583.0000 4081
##    Tumor_Sample_Barcode C&gt;G C&gt;T
## 1:         TCGA-A8-A08B   4   3
## 2:         TCGA-A8-A08B   1   8
## 3:         TCGA-A8-A08B   1   8
## 4:         TCGA-A8-A08B   1  21
## 5:         TCGA-A8-A08B   4   3
## 6:         TCGA-A8-A08B   4   4</code></pre>
<p><img src="maftools_files/figure-html/unnamed-chunk-12-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>“Kataegis” are defined as those genomic segments containing six or more consecutive mutations with an average inter-mutation distance of less than or equal to 1,00 bp <a href="#references">5</a>.</p>
</div>
<div id="compare-mutation-load-against-tcga-cohorts" class="section level2">
<h2 class="hasAnchor">
<a href="#compare-mutation-load-against-tcga-cohorts" class="anchor"></a>Compare mutation load against TCGA cohorts</h2>
<p><code>tcgaCompare</code> uses mutation load from TCGA <a href="https://gdc.cancer.gov/about-data/publications/mc3-2017">MC3</a> for comparing muttaion burden against 33 TCGA cohorts. Plot generated is <a href="http://www.nature.com/nature/journal/v500/n7463/fig_tab/nature12477_F1.html">similar</a> to the one described in Alexandrov et al <a href="#references">5</a>.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="no">laml.mutload</span> <span class="kw">=</span> <span class="fu"><a href="../reference/tcgaCompare.html">tcgaCompare</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">cohortName</span> <span class="kw">=</span> <span class="st">'Example-LAML'</span>)</pre></body></html></div>
<pre><code>## Warning in FUN(X[[i]], ...): Removed 1 samples with zero mutations.</code></pre>
<p><img src="maftools_files/figure-html/unnamed-chunk-13-1.png" width="1152" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="plotting-vaf" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-vaf" class="anchor"></a>Plotting VAF</h2>
<p>This function plots Variant Allele Frequencies as a boxplot which quickly helps to estimate clonal status of top mutated genes (clonal genes usually have mean allele frequency around ~50% assuming pure sample)</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="fu"><a href="../reference/plotVaf.html">plotVaf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">vafCol</span> <span class="kw">=</span> <span class="st">'i_TumorVAF_WU'</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-14-1.png" width="384" style="display: block; margin: auto auto auto 0;"></p>
</div>
</div>
<div id="processing-copy-number-data" class="section level1">
<h1 class="hasAnchor">
<a href="#processing-copy-number-data" class="anchor"></a>Processing copy-number data</h1>
<div id="reading-and-summarizing-gistic-output-files-" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-and-summarizing-gistic-output-files-" class="anchor"></a>Reading and summarizing gistic output files.</h2>
<p>We can summarize output files generated by GISTIC programme. As mentioned earlier, we need four files that are generated by GISTIC, i.e, all_lesions.conf_XX.txt, amp_genes.conf_XX.txt, del_genes.conf_XX.txt and scores.gistic, where XX is the confidence level. See <a href="ftp://ftp.broadinstitute.org/pub/GISTIC2.0/GISTICDocumentation_standalone.htm">GISTIC documentation</a> for details.</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="no">all.lesions</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"all_lesions.conf_99.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="no">amp.genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"amp_genes.conf_99.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="no">del.genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"del_genes.conf_99.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="no">scores.gis</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"scores.gistic"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)

<span class="no">laml.gistic</span> <span class="kw">=</span> <span class="fu"><a href="../reference/readGistic.html">readGistic</a></span>(<span class="kw">gisticAllLesionsFile</span> <span class="kw">=</span> <span class="no">all.lesions</span>, <span class="kw">gisticAmpGenesFile</span> <span class="kw">=</span> <span class="no">amp.genes</span>, <span class="kw">gisticDelGenesFile</span> <span class="kw">=</span> <span class="no">del.genes</span>, <span class="kw">gisticScoresFile</span> <span class="kw">=</span> <span class="no">scores.gis</span>, <span class="kw">isTCGA</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>## -Processing Gistic files..
## --Processing amp_genes.conf_99.txt
## --Processing del_genes.conf_99.txt
## --Processing scores.gistic
## --Summarizing by samples</code></pre>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="co">#GISTIC object</span>
<span class="no">laml.gistic</span></pre></body></html></div>
<pre><code>## An object of class  GISTIC 
##           ID summary
## 1:   Samples     191
## 2:    nGenes    2622
## 3: cytoBands      16
## 4:       Amp     388
## 5:       Del   26481
## 6:     total   26869</code></pre>
<p>Similar to MAF objects, there are methods available to access slots of GISTIC object - <code>getSampleSummary</code>, <code>getGeneSummary</code> and <code>getCytoBandSummary</code>. Summarized results can be written to output files using function <code>write.GisticSummary</code>.</p>
</div>
<div id="visualizing-gistic-results-" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-gistic-results-" class="anchor"></a>Visualizing gistic results.</h2>
<p>There are three types of plots available to visualize gistic results.</p>
<div id="genome-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#genome-plot" class="anchor"></a>genome plot</h3>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="fu"><a href="../reference/gisticChromPlot.html">gisticChromPlot</a></span>(<span class="kw">gistic</span> <span class="kw">=</span> <span class="no">laml.gistic</span>, <span class="kw">markBands</span> <span class="kw">=</span> <span class="st">"all"</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-16-1.png" width="576" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="bubble-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#bubble-plot" class="anchor"></a>Bubble plot</h3>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="fu"><a href="../reference/gisticBubblePlot.html">gisticBubblePlot</a></span>(<span class="kw">gistic</span> <span class="kw">=</span> <span class="no">laml.gistic</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-17-1.png" width="480" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="oncoplot" class="section level3">
<h3 class="hasAnchor">
<a href="#oncoplot" class="anchor"></a>oncoplot</h3>
<p>This is similar to oncoplots except for copy number data. One can again sort the matrix according to annotations, if any. Below plot is the gistic results for LAML, sorted according to FAB classification. Plot shows that 7q deletions are virtually absent in M4 subtype where as it is widespread in other subtypes.</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="fu"><a href="../reference/gisticOncoPlot.html">gisticOncoPlot</a></span>(<span class="kw">gistic</span> <span class="kw">=</span> <span class="no">laml.gistic</span>, <span class="kw">clinicalData</span> <span class="kw">=</span> <span class="fu"><a href="../reference/getClinicalData.html">getClinicalData</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">laml</span>), <span class="kw">clinicalFeatures</span> <span class="kw">=</span> <span class="st">'FAB_classification'</span>, <span class="kw">sortByAnnotation</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">top</span> <span class="kw">=</span> <span class="fl">10</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="visualizing-cbs-segments" class="section level3">
<h3 class="hasAnchor">
<a href="#visualizing-cbs-segments" class="anchor"></a>Visualizing CBS segments</h3>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="no">tcga.ab.009.seg</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"TCGA.AB.3009.hg19.seg.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="fu"><a href="../reference/plotCBSsegments.html">plotCBSsegments</a></span>(<span class="kw">cbsFile</span> <span class="kw">=</span> <span class="no">tcga.ab.009.seg</span>)</pre></body></html></div>
<pre><code>## No 'tsb' specified, plot head 1 sample. Set tsb='ALL' to plot all samples.</code></pre>
<pre><code>## NULL</code></pre>
<p><img src="maftools_files/figure-html/unnamed-chunk-19-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
<div id="analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis" class="anchor"></a>Analysis</h1>
<div id="somatic-interactions" class="section level2">
<h2 class="hasAnchor">
<a href="#somatic-interactions" class="anchor"></a>Somatic Interactions</h2>
<p>Mutually exclusive or co-occurring set of genes can be detected using <code>somaticInteractions</code> function, which performs pair-wise Fisher’s Exact test to detect such significant pair of genes.</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="co">#exclusive/co-occurance event analysis on top 10 mutated genes. </span>
<span class="fu"><a href="../reference/somaticInteractions.html">somaticInteractions</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">top</span> <span class="kw">=</span> <span class="fl">25</span>, <span class="kw">pvalue</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.1</span>))</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-20-1.png" width="480"></p>
<pre><code>##      gene1  gene2       pValue oddsRatio  00 11 01 10              Event
##   1: ASXL1  RUNX1 0.0001541586 55.215541 176  4 12  1       Co_Occurence
##   2:  IDH2  RUNX1 0.0002809928  9.590877 164  7  9 13       Co_Occurence
##   3:  IDH2  ASXL1 0.0004030636 41.077327 172  4  1 16       Co_Occurence
##   4:  FLT3   NPM1 0.0009929836  3.763161 125 17 16 35       Co_Occurence
##   5:  SMC3 DNMT3A 0.0010451985 20.177713 144  6 42  1       Co_Occurence
##  ---                                                                    
## 296: PLCE1  ASXL1 1.0000000000  0.000000 184  0  5  4 Mutually_Exclusive
## 297: RAD21  FAM5C 1.0000000000  0.000000 183  0  5  5 Mutually_Exclusive
## 298: PLCE1  FAM5C 1.0000000000  0.000000 184  0  5  4 Mutually_Exclusive
## 299: PLCE1  RAD21 1.0000000000  0.000000 184  0  5  4 Mutually_Exclusive
## 300:  EZH2  PLCE1 1.0000000000  0.000000 186  0  4  3 Mutually_Exclusive
##              pair event_ratio
##   1: ASXL1, RUNX1        4/13
##   2:  IDH2, RUNX1        7/22
##   3:  ASXL1, IDH2        4/17
##   4:   FLT3, NPM1       17/51
##   5: DNMT3A, SMC3        6/43
##  ---                         
## 296: ASXL1, PLCE1         0/9
## 297: FAM5C, RAD21        0/10
## 298: FAM5C, PLCE1         0/9
## 299: PLCE1, RAD21         0/9
## 300:  EZH2, PLCE1         0/7</code></pre>
</div>
<div id="detecting-cancer-driver-genes-based-on-positional-clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#detecting-cancer-driver-genes-based-on-positional-clustering" class="anchor"></a>Detecting cancer driver genes based on positional clustering</h2>
<p>maftools has a function <code>oncodrive</code> which identifies cancer genes (driver) from a given MAF. <code>oncodrive</code> is a based on algorithm <a href="http://bg.upf.edu/group/projects/oncodrive-clust.php">oncodriveCLUST</a> which was originally implemented in Python. Concept is based on the fact that most of the variants in cancer causing genes are enriched at few specific loci (aka hot-spots). This method takes advantage of such positions to identify cancer genes. If you use this function, please cite <a href="http://bioinformatics.oxfordjournals.org/content/early/2013/07/31/bioinformatics.btt395.full">OncodriveCLUST article</a> <a href="#references">7</a>.</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="no">laml.sig</span> <span class="kw">=</span> <span class="fu"><a href="../reference/oncodrive.html">oncodrive</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">AACol</span> <span class="kw">=</span> <span class="st">'Protein_Change'</span>, <span class="kw">minMut</span> <span class="kw">=</span> <span class="fl">5</span>, <span class="kw">pvalMethod</span> <span class="kw">=</span> <span class="st">'zscore'</span>)</pre></body></html></div>
<pre><code>## Warning in oncodrive(maf = laml, AACol = "Protein_Change", minMut = 5,
## pvalMethod = "zscore"): Oncodrive has been superseeded by OncodriveCLUSTL. See
## http://bg.upf.edu/group/projects/oncodrive-clust.php</code></pre>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">laml.sig</span>)</pre></body></html></div>
<pre><code>##    Hugo_Symbol Frame_Shift_Del Frame_Shift_Ins In_Frame_Del In_Frame_Ins
## 1:        IDH1               0               0            0            0
## 2:        IDH2               0               0            0            0
## 3:        NPM1               0              33            0            0
## 4:        NRAS               0               0            0            0
## 5:       U2AF1               0               0            0            0
## 6:         KIT               1               1            0            1
##    Missense_Mutation Nonsense_Mutation Splice_Site total MutatedSamples
## 1:                18                 0           0    18             18
## 2:                20                 0           0    20             20
## 3:                 1                 0           0    34             33
## 4:                15                 0           0    15             15
## 5:                 8                 0           0     8              8
## 6:                 7                 0           0    10              8
##    AlteredSamples clusters muts_in_clusters clusterScores protLen   zscore
## 1:             18        1               18     1.0000000     414 5.546154
## 2:             20        2               20     1.0000000     452 5.546154
## 3:             33        2               32     0.9411765     294 5.093665
## 4:             15        2               15     0.9218951     189 4.945347
## 5:              8        1                7     0.8750000     240 4.584615
## 6:              8        2                9     0.8500000     976 4.392308
##            pval          fdr fract_muts_in_clusters
## 1: 1.460110e-08 1.022077e-07              1.0000000
## 2: 1.460110e-08 1.022077e-07              1.0000000
## 3: 1.756034e-07 8.194826e-07              0.9411765
## 4: 3.800413e-07 1.330144e-06              1.0000000
## 5: 2.274114e-06 6.367520e-06              0.8750000
## 6: 5.607691e-06 1.308461e-05              0.9000000</code></pre>
<p>We can plot the results using <code>plotOncodrive</code>.</p>
<div class="sourceCode" id="cb46"><html><body><pre class="r"><span class="fu"><a href="../reference/plotOncodrive.html">plotOncodrive</a></span>(<span class="kw">res</span> <span class="kw">=</span> <span class="no">laml.sig</span>, <span class="kw">fdrCutOff</span> <span class="kw">=</span> <span class="fl">0.1</span>, <span class="kw">useFraction</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-23-1.png" width="480" style="display: block; margin: auto auto auto 0;"></p>
<p><code>plotOncodrive</code> plots the results as scatter plot with size of the points proportional to the number of clusters found in the gene. X-axis shows number of mutations (or fraction of mutations) observed in these clusters. In the above example, IDH1 has a single cluster and all of the 18 mutations are accumulated within that cluster, giving it a cluster score of one. For details on oncodrive algorithm, please refer to <a href="http://bioinformatics.oxfordjournals.org/content/early/2013/07/31/bioinformatics.btt395.full">OncodriveCLUST article</a> <a href="#references">7</a>.</p>
</div>
<div id="adding-and-summarizing-pfam-domains" class="section level2">
<h2 class="hasAnchor">
<a href="#adding-and-summarizing-pfam-domains" class="anchor"></a>Adding and summarizing pfam domains</h2>
<p>maftools comes with the function <code>pfamDomains</code>, which adds pfam domain information to the amino acid changes. <code>pfamDomain</code> also summarizes amino acid changes according to the domains that are affected. This serves the purpose of knowing what domain in given cancer cohort, is most frequently affected. This function is inspired from Pfam annotation module from MuSic tool <a href="#references">8</a>.</p>
<div class="sourceCode" id="cb47"><html><body><pre class="r"><span class="no">laml.pfam</span> <span class="kw">=</span> <span class="fu"><a href="../reference/pfamDomains.html">pfamDomains</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">AACol</span> <span class="kw">=</span> <span class="st">'Protein_Change'</span>, <span class="kw">top</span> <span class="kw">=</span> <span class="fl">10</span>)</pre></body></html></div>
<pre><code>## Warning in pfamDomains(maf = laml, AACol = "Protein_Change", top = 10): Removed
## 50 mutations for which AA position was not available</code></pre>
<p><img src="maftools_files/figure-html/unnamed-chunk-24-1.png" width="480" style="display: block; margin: auto auto auto 0;"></p>
<div class="sourceCode" id="cb49"><html><body><pre class="r"><span class="co">#Protein summary (Printing first 7 columns for display convenience)</span>
<span class="no">laml.pfam</span>$<span class="no">proteinSummary</span>[,<span class="fl">1</span>:<span class="fl">7</span>, <span class="kw">with</span> <span class="kw">=</span> <span class="fl">FALSE</span>]</pre></body></html></div>
<pre><code>##         HGNC AAPos Variant_Classification  N total  fraction   DomainLabel
##    1: DNMT3A   882      Missense_Mutation 27    54 0.5000000 AdoMet_MTases
##    2:   IDH1   132      Missense_Mutation 18    18 1.0000000      PTZ00435
##    3:   IDH2   140      Missense_Mutation 17    20 0.8500000      PTZ00435
##    4:   FLT3   835      Missense_Mutation 14    52 0.2692308      PKc_like
##    5:   FLT3   599           In_Frame_Ins 10    52 0.1923077      PKc_like
##   ---                                                                     
## 1512: ZNF646   875      Missense_Mutation  1     1 1.0000000          &lt;NA&gt;
## 1513: ZNF687   554      Missense_Mutation  1     2 0.5000000          &lt;NA&gt;
## 1514: ZNF687   363      Missense_Mutation  1     2 0.5000000          &lt;NA&gt;
## 1515: ZNF75D     5      Missense_Mutation  1     1 1.0000000          &lt;NA&gt;
## 1516: ZNF827   427        Frame_Shift_Del  1     1 1.0000000          &lt;NA&gt;</code></pre>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="co">#Domain summary (Printing first 3 columns for display convenience)</span>
<span class="no">laml.pfam</span>$<span class="no">domainSummary</span>[,<span class="fl">1</span>:<span class="fl">3</span>, <span class="kw">with</span> <span class="kw">=</span> <span class="fl">FALSE</span>]</pre></body></html></div>
<pre><code>##        DomainLabel nMuts nGenes
##   1:      PKc_like    55      5
##   2:      PTZ00435    38      2
##   3: AdoMet_MTases    33      1
##   4:         7tm_1    24     24
##   5:       COG5048    17     17
##  ---                           
## 499:    ribokinase     1      1
## 500:   rim_protein     1      1
## 501: sigpep_I_bact     1      1
## 502:           trp     1      1
## 503:        zf-BED     1      1</code></pre>
</div>
<div id="survival-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#survival-analysis" class="anchor"></a>Survival analysis</h2>
<p>Survival analysis is an essential part of cohort based sequencing projects. Function <code>mafSurvive</code> performs survival analysis and draws kaplan meier curve by grouping samples based on mutation status of user defined gene(s) or manually provided samples those make up a group. This function requires input data to contain Tumor_Sample_Barcode (make sure they match to those in MAF file), binary event (1/0) and time to event.</p>
<p>Our annotation data already contains survival information and in case you have survival data stored in a separate table provide them via argument <code>clinicalData</code></p>
<div id="mutation-in-any-given-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#mutation-in-any-given-genes" class="anchor"></a>Mutation in any given genes</h3>
<div class="sourceCode" id="cb53"><html><body><pre class="r"><span class="co">#Survival analysis based on grouping of DNMT3A mutation status</span>
<span class="fu"><a href="../reference/mafSurvival.html">mafSurvival</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">genes</span> <span class="kw">=</span> <span class="st">'DNMT3A'</span>, <span class="kw">time</span> <span class="kw">=</span> <span class="st">'days_to_last_followup'</span>, <span class="kw">Status</span> <span class="kw">=</span> <span class="st">'Overall_Survival_Status'</span>, <span class="kw">isTCGA</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>## Looking for clinical data in annoatation slot of MAF..</code></pre>
<pre><code>## Number of mutated samples for given genes:</code></pre>
<pre><code>## DNMT3A 
##     48</code></pre>
<pre><code>## Removed 11 samples with NA's</code></pre>
<pre><code>## Median survival..</code></pre>
<pre><code>##     Group medianTime   N
## 1: Mutant        245  45
## 2:     WT        396 137</code></pre>
<p><img src="maftools_files/figure-html/unnamed-chunk-25-1.png" width="480"></p>
</div>
<div id="predict-genesets-associated-with-survival" class="section level3">
<h3 class="hasAnchor">
<a href="#predict-genesets-associated-with-survival" class="anchor"></a>Predict genesets associated with survival</h3>
<p>Identify set of genes which results in poor survival</p>
<div class="sourceCode" id="cb60"><html><body><pre class="r"><span class="co">#Using top 20 mutated genes to identify a set of genes (of size 2) to predict poor prognostic groups</span>
<span class="no">prog_geneset</span> <span class="kw">=</span> <span class="fu"><a href="../reference/survGroup.html">survGroup</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">top</span> <span class="kw">=</span> <span class="fl">20</span>, <span class="kw">geneSetSize</span> <span class="kw">=</span> <span class="fl">2</span>, <span class="kw">time</span> <span class="kw">=</span> <span class="st">"days_to_last_followup"</span>, <span class="kw">Status</span> <span class="kw">=</span> <span class="st">"Overall_Survival_Status"</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<pre><code>## Removed 11 samples with NA's</code></pre>
<div class="sourceCode" id="cb62"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">prog_geneset</span>)</pre></body></html></div>
<pre><code>##     Gene_combination P_value    hr  WT Mutant
##  1:      FLT3_DNMT3A 0.00104 2.510 164     18
##  2:      DNMT3A_SMC3 0.04880 2.220 176      6
##  3:      DNMT3A_NPM1 0.07190 1.720 166     16
##  4:      DNMT3A_TET2 0.19600 1.780 176      6
##  5:        FLT3_TET2 0.20700 1.860 177      5
##  6:        NPM1_IDH1 0.21900 0.495 176      6
##  7:      DNMT3A_IDH1 0.29300 1.500 173      9
##  8:       IDH2_RUNX1 0.31800 1.580 176      6
##  9:        FLT3_NPM1 0.53600 1.210 165     17
## 10:      DNMT3A_IDH2 0.68000 0.747 178      4
## 11:      DNMT3A_NRAS 0.99200 0.986 178      4</code></pre>
<p>Above results show a combination (N = 2) of genes which are associated with poor survival (P &lt; 0.05). We can draw KM curve for above results with the function <code>mafSurvGroup</code></p>
<div class="sourceCode" id="cb64"><html><body><pre class="r"><span class="fu"><a href="../reference/mafSurvGroup.html">mafSurvGroup</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">geneSet</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"DNMT3A"</span>, <span class="st">"FLT3"</span>), <span class="kw">time</span> <span class="kw">=</span> <span class="st">"days_to_last_followup"</span>, <span class="kw">Status</span> <span class="kw">=</span> <span class="st">"Overall_Survival_Status"</span>)</pre></body></html></div>
<pre><code>## Looking for clinical data in annoatation slot of MAF..</code></pre>
<pre><code>## Removed 11 samples with NA's</code></pre>
<pre><code>## Median survival..</code></pre>
<pre><code>##     Group medianTime   N
## 1: Mutant      242.5  18
## 2:     WT      379.5 164</code></pre>
<p><img src="maftools_files/figure-html/unnamed-chunk-27-1.png" width="480"></p>
</div>
</div>
<div id="comparing-two-cohorts-mafs" class="section level2">
<h2 class="hasAnchor">
<a href="#comparing-two-cohorts-mafs" class="anchor"></a>Comparing two cohorts (MAFs)</h2>
<p>Cancers differ from each other in terms of their mutation pattern. We can compare two different cohorts to detect such differentially mutated genes. For example, recent article by <a href="http://www.ncbi.nlm.nih.gov/pubmed/27063598">Madan et. al</a> <a href="references">9</a>, have shown that patients with relapsed APL (Acute Promyelocytic Leukemia) tends to have mutations in PML and RARA genes, which were absent during primary stage of the disease. This difference between two cohorts (in this case primary and relapse APL) can be detected using function <code>mafComapre</code>, which performs fisher test on all genes between two cohorts to detect differentially mutated genes.</p>
<div class="sourceCode" id="cb69"><html><body><pre class="r"><span class="co">#Primary APL MAF</span>
<span class="no">primary.apl</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"APL_primary.maf.gz"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="no">primary.apl</span> <span class="kw">=</span> <span class="fu"><a href="../reference/read.maf.html">read.maf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">primary.apl</span>)
<span class="co">#Relapse APL MAF</span>
<span class="no">relapse.apl</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"APL_relapse.maf.gz"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="no">relapse.apl</span> <span class="kw">=</span> <span class="fu"><a href="../reference/read.maf.html">read.maf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">relapse.apl</span>)</pre></body></html></div>
<div class="sourceCode" id="cb70"><html><body><pre class="r"><span class="co">#Considering only genes which are mutated in at-least in 5 samples in one of the cohort to avoid bias due to genes mutated in single sample.</span>
<span class="no">pt.vs.rt</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/mafCompare.html">mafCompare</a></span>(<span class="kw">m1</span> <span class="kw">=</span> <span class="no">primary.apl</span>, <span class="kw">m2</span> <span class="kw">=</span> <span class="no">relapse.apl</span>, <span class="kw">m1Name</span> <span class="kw">=</span> <span class="st">'Primary'</span>, <span class="kw">m2Name</span> <span class="kw">=</span> <span class="st">'Relapse'</span>, <span class="kw">minMut</span> <span class="kw">=</span> <span class="fl">5</span>)
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">pt.vs.rt</span>)</pre></body></html></div>
<pre><code>## $results
##    Hugo_Symbol Primary Relapse         pval         or       ci.up      ci.low
## 1:         PML       1      11 1.529935e-05 0.03537381   0.2552937 0.000806034
## 2:        RARA       0       7 2.574810e-04 0.00000000   0.3006159 0.000000000
## 3:       RUNX1       1       5 1.310500e-02 0.08740567   0.8076265 0.001813280
## 4:        FLT3      26       4 1.812779e-02 3.56086275  14.7701728 1.149009169
## 5:      ARID1B       5       8 2.758396e-02 0.26480490   0.9698686 0.064804160
## 6:         WT1      20      14 2.229087e-01 0.60619329   1.4223101 0.263440988
## 7:        KRAS       6       1 4.334067e-01 2.88486293 135.5393108 0.337679367
## 8:        NRAS      15       4 4.353567e-01 1.85209500   8.0373994 0.553883512
## 9:      ARID1A       7       4 7.457274e-01 0.80869223   3.9297309 0.195710173
##         adjPval
## 1: 0.0001376942
## 2: 0.0011586643
## 3: 0.0393149868
## 4: 0.0407875250
## 5: 0.0496511201
## 6: 0.3343630535
## 7: 0.4897762916
## 8: 0.4897762916
## 9: 0.7457273717
## 
## $SampleSummary
##     Cohort SampleSize
## 1: Primary        124
## 2: Relapse         58</code></pre>
<div id="forest-plots" class="section level3">
<h3 class="hasAnchor">
<a href="#forest-plots" class="anchor"></a>Forest plots</h3>
<p>Above results show two genes PML and RARA which are highly mutated in Relapse APL compared to Primary APL. We can visualize these results as a <a href="https://en.wikipedia.org/wiki/Forest_plot">forestplot</a>.</p>
<div class="sourceCode" id="cb72"><html><body><pre class="r"><span class="fu"><a href="../reference/forestPlot.html">forestPlot</a></span>(<span class="kw">mafCompareRes</span> <span class="kw">=</span> <span class="no">pt.vs.rt</span>, <span class="kw">pVal</span> <span class="kw">=</span> <span class="fl">0.1</span>, <span class="kw">color</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'royalblue'</span>, <span class="st">'maroon'</span>), <span class="kw">geneFontSize</span> <span class="kw">=</span> <span class="fl">0.8</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-30-1.png" width="480" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="co-onco-plots" class="section level3">
<h3 class="hasAnchor">
<a href="#co-onco-plots" class="anchor"></a>Co-onco plots</h3>
<p>Another alternative way of displaying above results is by plotting two oncoplots side by side. <code>coOncoplot</code> function takes two maf objects and plots them side by side for better comparison.</p>
<div class="sourceCode" id="cb73"><html><body><pre class="r"><span class="no">genes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"PML"</span>, <span class="st">"RARA"</span>, <span class="st">"RUNX1"</span>, <span class="st">"ARID1B"</span>, <span class="st">"FLT3"</span>)
<span class="fu"><a href="../reference/coOncoplot.html">coOncoplot</a></span>(<span class="kw">m1</span> <span class="kw">=</span> <span class="no">primary.apl</span>, <span class="kw">m2</span> <span class="kw">=</span> <span class="no">relapse.apl</span>, <span class="kw">m1Name</span> <span class="kw">=</span> <span class="st">'PrimaryAPL'</span>, <span class="kw">m2Name</span> <span class="kw">=</span> <span class="st">'RelapseAPL'</span>, <span class="kw">genes</span> <span class="kw">=</span> <span class="no">genes</span>, <span class="kw">removeNonMutated</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-31-1.png" width="1056" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="co-bar-plots" class="section level3">
<h3 class="hasAnchor">
<a href="#co-bar-plots" class="anchor"></a>Co-bar plots</h3>
<div class="sourceCode" id="cb74"><html><body><pre class="r"><span class="fu"><a href="../reference/coBarplot.html">coBarplot</a></span>(<span class="kw">m1</span> <span class="kw">=</span> <span class="no">primary.apl</span>, <span class="kw">m2</span> <span class="kw">=</span> <span class="no">relapse.apl</span>, <span class="kw">m1Name</span> <span class="kw">=</span> <span class="st">"Primary"</span>, <span class="kw">m2Name</span> <span class="kw">=</span> <span class="st">"Relapse"</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-32-1.png" width="480"></p>
</div>
<div id="lollipop-plot-2" class="section level3">
<h3 class="hasAnchor">
<a href="#lollipop-plot-2" class="anchor"></a>Lollipop plot-2</h3>
<p>Along with plots showing cohort wise differences, its also possible to show gene wise differences with <code>lollipopPlot2</code> function.</p>
<div class="sourceCode" id="cb75"><html><body><pre class="r"><span class="fu"><a href="../reference/lollipopPlot2.html">lollipopPlot2</a></span>(<span class="kw">m1</span> <span class="kw">=</span> <span class="no">primary.apl</span>, <span class="kw">m2</span> <span class="kw">=</span> <span class="no">relapse.apl</span>, <span class="kw">gene</span> <span class="kw">=</span> <span class="st">"PML"</span>, <span class="kw">AACol1</span> <span class="kw">=</span> <span class="st">"amino_acid_change"</span>, <span class="kw">AACol2</span> <span class="kw">=</span> <span class="st">"amino_acid_change"</span>, <span class="kw">m1_name</span> <span class="kw">=</span> <span class="st">"Primary"</span>, <span class="kw">m2_name</span> <span class="kw">=</span> <span class="st">"Relapse"</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-33-1.png" width="700" style="display: block; margin: auto auto auto 0;"></p>
</div>
</div>
<div id="clinical-enrichment-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#clinical-enrichment-analysis" class="anchor"></a>Clinical enrichment analysis</h2>
<p><code>clinicalEnrichment</code> is another function which takes any clinical feature associated with the samples and performs enrichment analysis. It performs various groupwise and pairwise comparisions to identify enriched mutations for every category within a clincila feature. Below is an example to identify mutations associated with FAB_classification.</p>
<div class="sourceCode" id="cb76"><html><body><pre class="r"><span class="no">fab.ce</span> <span class="kw">=</span> <span class="fu"><a href="../reference/clinicalEnrichment.html">clinicalEnrichment</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">clinicalFeature</span> <span class="kw">=</span> <span class="st">'FAB_classification'</span>)</pre></body></html></div>
<pre><code>## Sample size per factor in FAB_classification:</code></pre>
<pre><code>## 
## M0 M1 M2 M3 M4 M5 M6 M7 
## 19 44 44 21 39 19  3  3</code></pre>
<div class="sourceCode" id="cb79"><html><body><pre class="r"><span class="co">#Results are returned as a list. Significant associations p-value &lt; 0.05</span>
<span class="no">fab.ce</span>$<span class="no">groupwise_comparision</span>[<span class="no">p_value</span> <span class="kw">&lt;</span> <span class="fl">0.05</span>]</pre></body></html></div>
<pre><code>##    Hugo_Symbol Group1 Group2 n_mutated_group1 n_mutated_group2      p_value
## 1:        IDH1     M1   Rest         11 of 44         7 of 149 0.0002597371
## 2:        TP53     M7   Rest           3 of 3        12 of 190 0.0003857187
## 3:      DNMT3A     M5   Rest         10 of 19        38 of 174 0.0057610493
## 4:       CEBPA     M2   Rest          7 of 44         6 of 149 0.0117352110
## 5:       RUNX1     M0   Rest          5 of 19        11 of 174 0.0117436825
## 6:        NPM1     M5   Rest          7 of 19        26 of 174 0.0248582372
## 7:       CEBPA     M1   Rest          6 of 44         7 of 149 0.0478737468
##    OR_low   OR_high       fdr
## 1:      0 0.3926994 0.0308575
## 2:      0 0.1315271 0.0308575
## 3:      0 0.6406007 0.3072560
## 4:      0 0.6874270 0.3757978
## 5:      0 0.6466787 0.3757978
## 6:      0 0.8342897 0.6628863
## 7:      0 0.9869971 1.0000000</code></pre>
<p>Above results shows IDH1 mutations are enriched in M1 subtype of leukemia compared to rest of the cohort. Similarly DNMT3A is in M5, RUNX1 is in M0, and so on. These are well known results and this function effectively recaptures them. One can use any sort of clincial feature to perform such an analysis. There is also a small function - <code>plotEnrichmentResults</code> which can be used to plot these results.</p>
<div class="sourceCode" id="cb81"><html><body><pre class="r"><span class="fu"><a href="../reference/plotEnrichmentResults.html">plotEnrichmentResults</a></span>(<span class="kw">enrich_res</span> <span class="kw">=</span> <span class="no">fab.ce</span>, <span class="kw">pVal</span> <span class="kw">=</span> <span class="fl">0.05</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-35-1.png" width="576"></p>
</div>
<div id="drug-gene-interactions" class="section level2">
<h2 class="hasAnchor">
<a href="#drug-gene-interactions" class="anchor"></a>Drug-Gene Interactions</h2>
<p><code>drugInteractions</code> function checks for drug–gene interactions and gene druggability information compiled from <a href="http://www.dgidb.org">Drug Gene Interaction database</a>.</p>
<div class="sourceCode" id="cb82"><html><body><pre class="r"><span class="no">dgi</span> <span class="kw">=</span> <span class="fu"><a href="../reference/drugInteractions.html">drugInteractions</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">fontSize</span> <span class="kw">=</span> <span class="fl">0.75</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-36-1.png" width="768"></p>
<p>Above plot shows potential druggable gene categories along with upto top 5 genes involved in them. One can also extract information on drug-gene interactions. For example below is the results for known/reported drugs to interact with DNMT3A.</p>
<div class="sourceCode" id="cb83"><html><body><pre class="r"><span class="no">dnmt3a.dgi</span> <span class="kw">=</span> <span class="fu"><a href="../reference/drugInteractions.html">drugInteractions</a></span>(<span class="kw">genes</span> <span class="kw">=</span> <span class="st">"DNMT3A"</span>, <span class="kw">drugs</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>## Number of claimed drugs for given genes:
##      Gene N
## 1: DNMT3A 7</code></pre>
<div class="sourceCode" id="cb85"><html><body><pre class="r"><span class="co">#Printing selected columns.</span>
<span class="no">dnmt3a.dgi</span>[,<span class="fu">.</span>(<span class="no">Gene</span>, <span class="no">interaction_types</span>, <span class="no">drug_name</span>, <span class="no">drug_claim_name</span>)]</pre></body></html></div>
<pre><code>##      Gene interaction_types    drug_name drug_claim_name
## 1: DNMT3A                                            N/A
## 2: DNMT3A                   DAUNORUBICIN    Daunorubicin
## 3: DNMT3A                     DECITABINE      Decitabine
## 4: DNMT3A                     IDARUBICIN      IDARUBICIN
## 5: DNMT3A                     DECITABINE      DECITABINE
## 6: DNMT3A         inhibitor   DECITABINE   CHEMBL1201129
## 7: DNMT3A         inhibitor  AZACITIDINE      CHEMBL1489</code></pre>
<p>Please cite <a href="https://www.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;retmode=ref&amp;cmd=prlinks&amp;id=24122041">DGIdb article</a> if you find this function useful <a href="#references">10</a>.</p>
<p><em>Disclaimer: Resources used in this function are intended for purely research purposes. It should not be used for emergencies or medical or professional advice.</em></p>
</div>
<div id="oncogenic-signaling-pathways" class="section level2">
<h2 class="hasAnchor">
<a href="#oncogenic-signaling-pathways" class="anchor"></a>Oncogenic Signaling Pathways</h2>
<p><code>OncogenicPathways</code> function checks for enrichment of known Oncogenic Signaling Pathways in TCGA cohorts <a href="#references">11</a>.</p>
<div class="sourceCode" id="cb87"><html><body><pre class="r"><span class="fu"><a href="../reference/OncogenicPathways.html">OncogenicPathways</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-38-1.png" width="384"></p>
<pre><code>##    Pathway  N n_affected_genes fraction_affected Mutated_samples
## 1:    PI3K 29                1        0.03448276               1
## 2:    NRF2  3                1        0.33333333               1
## 3:    TP53  6                2        0.33333333              15
## 4:     WNT 68                3        0.04411765               4
## 5:     MYC 13                3        0.23076923               3
## 6:   NOTCH 71                6        0.08450704               8
## 7:   Hippo 38                7        0.18421053               7
## 8: RTK-RAS 85               18        0.21176471              97
##    Fraction_mutated_samples
## 1:              0.005181347
## 2:              0.005181347
## 3:              0.077720207
## 4:              0.020725389
## 5:              0.015544041
## 6:              0.041450777
## 7:              0.036269430
## 8:              0.502590674</code></pre>
<p>Its also possible to visualize complete pathway.</p>
<div class="sourceCode" id="cb89"><html><body><pre class="r"><span class="fu"><a href="../reference/PlotOncogenicPathways.html">PlotOncogenicPathways</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">pathways</span> <span class="kw">=</span> <span class="st">"RTK-RAS"</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-39-1.png" width="960"></p>
<p>Tumor suppressor genes are in red, and oncogenes are in blue font.</p>
</div>
<div id="tumor-heterogeneity-and-math-scores" class="section level2">
<h2 class="hasAnchor">
<a href="#tumor-heterogeneity-and-math-scores" class="anchor"></a>Tumor heterogeneity and MATH scores</h2>
<div id="heterogeneity-in-tumor-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#heterogeneity-in-tumor-samples" class="anchor"></a>Heterogeneity in tumor samples</h3>
<p>Tumors are generally heterogeneous i.e, consist of multiple clones. This heterogeneity can be inferred by clustering variant allele frequencies. <code>inferHeterogeneity</code> function uses vaf information to cluster variants (using <code>mclust</code>), to infer clonality. By default, <code>inferHeterogeneity</code> function looks for column <em>t_vaf</em> containing vaf information. However, if the field name is different from <em>t_vaf</em>, we can manually specify it using argument <code>vafCol</code>. For example, in this case study vaf is stored under the field name <em>i_TumorVAF_WU</em>.</p>
<div class="sourceCode" id="cb90"><html><body><pre class="r"><span class="co">#Heterogeneity in sample TCGA.AB.2972</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"mclust"</span>)</pre></body></html></div>
<pre><code>## Package 'mclust' version 5.4.6
## Type 'citation("mclust")' for citing this R package in publications.</code></pre>
<div class="sourceCode" id="cb92"><html><body><pre class="r"><span class="no">tcga.ab.2972.het</span> <span class="kw">=</span> <span class="fu"><a href="../reference/inferHeterogeneity.html">inferHeterogeneity</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">tsb</span> <span class="kw">=</span> <span class="st">'TCGA-AB-2972'</span>, <span class="kw">vafCol</span> <span class="kw">=</span> <span class="st">'i_TumorVAF_WU'</span>)</pre></body></html></div>
<pre><code>## Processing TCGA-AB-2972..</code></pre>
<div class="sourceCode" id="cb94"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">tcga.ab.2972.het</span>$<span class="no">clusterMeans</span>)</pre></body></html></div>
<pre><code>##    Tumor_Sample_Barcode cluster   meanVaf
## 1:         TCGA-AB-2972       2 0.4496571
## 2:         TCGA-AB-2972       1 0.2454750
## 3:         TCGA-AB-2972 outlier 0.3695000</code></pre>
<div class="sourceCode" id="cb96"><html><body><pre class="r"><span class="co">#Visualizing results</span>
<span class="fu"><a href="../reference/plotClusters.html">plotClusters</a></span>(<span class="kw">clusters</span> <span class="kw">=</span> <span class="no">tcga.ab.2972.het</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-40-1.png" width="576" style="display: block; margin: auto auto auto 0;"></p>
<p>Above figure shows clear separation of two clones clustered at mean variant allele frequencies of ~45% (major clone) and another minor clone at variant allele frequency of ~25%.</p>
<p>Although clustering of variant allele frequencies gives us a fair idea on heterogeneity, it is also possible to measure the extent of heterogeneity in terms of a numerical value. MATH score (mentioned as a subtitle in above plot) is a simple quantitative measure of intra-tumor heterogeneity, which calculates the width of the vaf distribution. Higher MATH scores are found to be associated with poor outcome. MATH score can also be used a proxy variable for survival analysis <a href="#references">11</a>.</p>
</div>
<div id="ignoring-variants-in-copy-number-altered-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#ignoring-variants-in-copy-number-altered-regions" class="anchor"></a>Ignoring variants in copy number altered regions</h3>
<p>We can use copy number information to ignore variants located on copy-number altered regions. Copy number alterations results in abnormally high/low variant allele frequencies, which tends to affect clustering. Removing such variants improves clustering and density estimation while retaining biologically meaningful results. Copy number information can be provided as a segmented file generated from segmentation programmes, such as Circular Binary Segmentation from “DNACopy” Bioconductor package <a href="#references">6</a>.</p>
<div class="sourceCode" id="cb97"><html><body><pre class="r"><span class="no">seg</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'TCGA.AB.3009.hg19.seg.txt'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'maftools'</span>)
<span class="no">tcga.ab.3009.het</span> <span class="kw">=</span> <span class="fu"><a href="../reference/inferHeterogeneity.html">inferHeterogeneity</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">tsb</span> <span class="kw">=</span> <span class="st">'TCGA-AB-3009'</span>, <span class="kw">segFile</span> <span class="kw">=</span> <span class="no">seg</span>, <span class="kw">vafCol</span> <span class="kw">=</span> <span class="st">'i_TumorVAF_WU'</span>)</pre></body></html></div>
<pre><code>## Processing TCGA-AB-3009..</code></pre>
<pre><code>## Removed 1 variants with no copy number data.</code></pre>
<pre><code>##    Hugo_Symbol Chromosome Start_Position End_Position Tumor_Sample_Barcode
## 1:        PHF6         23      133551224    133551224         TCGA-AB-3009
##        t_vaf Segment_Start Segment_End Segment_Mean CN
## 1: 0.9349112            NA          NA           NA NA</code></pre>
<pre><code>## Copy number altered variants:</code></pre>
<pre><code>##    Hugo_Symbol Chromosome Start_Position End_Position Tumor_Sample_Barcode
## 1:     NFKBIL2          8      145668658    145668658         TCGA-AB-3009
## 2:         NF1         17       29562981     29562981         TCGA-AB-3009
## 3:       SUZ12         17       30293198     30293198         TCGA-AB-3009
##        t_vaf Segment_Start Segment_End Segment_Mean       CN    cluster
## 1: 0.4415584     145232496   145760746       0.3976 2.634629 CN_altered
## 2: 0.8419000      29054355    30363868      -0.9157 1.060173 CN_altered
## 3: 0.8958333      29054355    30363868      -0.9157 1.060173 CN_altered</code></pre>
<div class="sourceCode" id="cb103"><html><body><pre class="r"><span class="co">#Visualizing results. Highlighting those variants on copynumber altered variants.</span>
<span class="fu"><a href="../reference/plotClusters.html">plotClusters</a></span>(<span class="kw">clusters</span> <span class="kw">=</span> <span class="no">tcga.ab.3009.het</span>, <span class="kw">genes</span> <span class="kw">=</span> <span class="st">'CN_altered'</span>, <span class="kw">showCNvars</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-41-1.png" width="576" style="display: block; margin: auto auto auto 0;"></p>
<p>Above figure shows two genes NF1 and SUZ12 with high VAF’s, which is due to copy number alterations (deletion). Those two genes are ignored from analysis.</p>
</div>
</div>
<div id="mutational-signatures" class="section level2">
<h2 class="hasAnchor">
<a href="#mutational-signatures" class="anchor"></a>Mutational Signatures</h2>
<p>Every cancer, as it progresses leaves a signature characterized by specific pattern of nucleotide substitutions. <a href="http://www.nature.com/nature/journal/v500/n7463/full/nature12477.html">Alexandrov et.al</a> have shown such mutational signatures, derived from over 7000 cancer samples <a href="#references">5</a>. Such signatures can be extracted by decomposing matrix of nucleotide substitutions, classified into 96 substitution classes based on immediate bases surrounding the mutated base. Extracted signatures can also be compared to those <a href="http://cancer.sanger.ac.uk/cosmic/signatures">validated signatures</a>.</p>
<p>First step in signature analysis is to obtain the adjacent bases surrounding the mutated base and form a mutation matrix. NOTE: Earlier versions of maftools required a fasta file as an input. But starting from 1.8.0, BSgenome objects are used for faster sequence extraction.</p>
<div class="sourceCode" id="cb104"><html><body><pre class="r"><span class="co">#Requires BSgenome object</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BSgenome.Hsapiens.UCSC.hg19</span>, <span class="kw">quietly</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>## 
## Attaching package: 'BiocGenerics'</code></pre>
<pre><code>## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which, which.max, which.min</code></pre>
<pre><code>## 
## Attaching package: 'S4Vectors'</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     expand.grid</code></pre>
<pre><code>## 
## Attaching package: 'Biostrings'</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     strsplit</code></pre>
<div class="sourceCode" id="cb113"><html><body><pre class="r"><span class="no">laml.tnm</span> <span class="kw">=</span> <span class="fu"><a href="../reference/trinucleotideMatrix.html">trinucleotideMatrix</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">prefix</span> <span class="kw">=</span> <span class="st">'chr'</span>, <span class="kw">add</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">ref_genome</span> <span class="kw">=</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>)</pre></body></html></div>
<pre><code>## Warning in trinucleotideMatrix(maf = laml, prefix = "chr", add = TRUE, ref_genome = "BSgenome.Hsapiens.UCSC.hg19"): Chromosome names in MAF must match chromosome names in reference genome.
## Ignorinig 101 single nucleotide variants from missing chromosomes chr23</code></pre>
<pre><code>## -Extracting 5' and 3' adjacent bases
## -Extracting +/- 20bp around mutated bases for background C&gt;T estimation
## -Estimating APOBEC enrichment scores
## --Performing one-way Fisher's test for APOBEC enrichment
## ---APOBEC related mutations are enriched in  3.315 % of samples (APOBEC enrichment score &gt; 2 ;  6  of  181  samples)
## -Creating mutation matrix
## --matrix of dimension 188x96</code></pre>
<p>Above function performs two steps:</p>
<ul>
<li>Estimates APOBEC enrichment scores</li>
<li>Prepares a mutational matrix for signature analysis.</li>
</ul>
<div id="apobec-enrichment-estimation-" class="section level3">
<h3 class="hasAnchor">
<a href="#apobec-enrichment-estimation-" class="anchor"></a>APOBEC Enrichment estimation.</h3>
<p>APOBEC induced mutations are more frequent in solid tumors and are mainly associated with C&gt;T transition events occurring in TCW motif. APOBEC enrichment scores in the above command are estimated using the method described by Roberts et al <a href="#references">13</a>. Briefly, enrichment of C&gt;T mutations occurring within TCW motif over all of the C&gt;T mutations in a given sample is compared to background Cytosines and TCWs occurring within 20bp of mutated bases.</p>
<p><span class="math display">\[\frac{n_{tCw} * background_C}{n_C * background_{TCW}}\]</span></p>
<p>One-sided fishers exact test is also performed to statistically evaluate the enrichment score, as described in original study by Roberts et al.</p>
</div>
<div id="differences-between-apobec-enriched-and-non-enriched-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#differences-between-apobec-enriched-and-non-enriched-samples" class="anchor"></a>Differences between APOBEC enriched and non-enriched samples</h3>
<p>We can also analyze the differences in mutational patterns between APOBEC enriched and non-APOBEC enriched samples. <code>plotApobecDiff</code> is a function which takes APOBEC enrichment scores estimated by <code>trinucleotideMatrix</code> and classifies samples into APOBEC enriched and non-APOBEC enriched. Once stratified, it compares these two groups to identify differentially altered genes.</p>
<p>Note that, LAML with no APOBEC enrichments, is not an ideal cohort for this sort of analysis and hence below plot is only for demonstration purpose.</p>
<div class="sourceCode" id="cb116"><html><body><pre class="r"><span class="fu"><a href="../reference/plotApobecDiff.html">plotApobecDiff</a></span>(<span class="kw">tnm</span> <span class="kw">=</span> <span class="no">laml.tnm</span>, <span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">pVal</span> <span class="kw">=</span> <span class="fl">0.2</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-43-1.png" width="672"></p>
<pre><code>## $results
##      Hugo_Symbol Enriched nonEnriched       pval        or      ci.up
##   1:        TP53        2          13 0.08175632 5.9976455  46.608861
##   2:        TET2        1          16 0.45739351 1.9407002  18.983979
##   3:        FLT3        2          45 0.65523131 1.4081851  10.211621
##   4:      DNMT3A        1          47 1.00000000 0.5335362   4.949499
##   5:      ADAM11        0           2 1.00000000 0.0000000 164.191472
##  ---                                                                 
## 132:         WAC        0           2 1.00000000 0.0000000 164.191472
## 133:         WT1        0          12 1.00000000 0.0000000  12.690862
## 134:      ZBTB33        0           2 1.00000000 0.0000000 164.191472
## 135:      ZC3H18        0           2 1.00000000 0.0000000 164.191472
## 136:      ZNF687        0           2 1.00000000 0.0000000 164.191472
##          ci.low adjPval
##   1: 0.49875432       1
##   2: 0.03882963       1
##   3: 0.12341748       1
##   4: 0.01101929       1
##   5: 0.00000000       1
##  ---                   
## 132: 0.00000000       1
## 133: 0.00000000       1
## 134: 0.00000000       1
## 135: 0.00000000       1
## 136: 0.00000000       1
## 
## $SampleSummary
##         Cohort SampleSize  Mean Median
## 1:    Enriched          6 7.167    6.5
## 2: nonEnriched        172 9.715    9.0</code></pre>
</div>
<div id="signature-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#signature-analysis" class="anchor"></a>Signature analysis</h3>
<p>Signature analysis includes following steps.</p>
<ol style="list-style-type: decimal">
<li>
<code>estimateSignatures</code> - which runs NMF on a range of values and measures the goodness of fit - in terms of <a href="https://en.wikipedia.org/wiki/Cophenetic_correlation">Cophenetic correlation</a>.</li>
<li>
<code>plotCophenetic</code> - which draws an elblow plot and helps you to decide optimal number of signatures. Best possible signature is the value at which Cophenetic correlation drops significantly.</li>
<li>
<code>extractSignatures</code> - uses non-negative matrix factorization to decompose the matrix into <code>n</code> signatures. <code>n</code> is chosen based on the above two steps. In case if you already have a good estimate of <code>n</code>, you can skip above two steps.</li>
<li>
<code>compareSignatures</code> - extracted signatures from above step can be compared to known signatures<a href="#references">11</a> from <a href="https://cancer.sanger.ac.uk/cosmic/signatures/SBS/">COSMIC</a> database, and cosine similarity is calculated to identify best match.</li>
<li>
<code>plotSignatures</code> - plots signatures</li>
</ol>
<p><img src="maftools_files/figure-html/unnamed-chunk-44-1.png" width="700"></p>
<p><strong><em>Note:</em></strong> In previous versions, <code>extractSignatures</code> used to take care of above steps automatically. After versions 2.2.0, main function is split inot above 5 stpes for user flexibility.</p>
<div class="sourceCode" id="cb118"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">'NMF'</span>)
<span class="no">laml.sign</span> <span class="kw">=</span> <span class="fu"><a href="../reference/estimateSignatures.html">estimateSignatures</a></span>(<span class="kw">mat</span> <span class="kw">=</span> <span class="no">laml.tnm</span>, <span class="kw">nTry</span> <span class="kw">=</span> <span class="fl">6</span>)</pre></body></html></div>
<p>Draw elbow plot to visualize and decide optimal number of signatures from above results.</p>
<div class="sourceCode" id="cb119"><html><body><pre class="r"><span class="fu"><a href="../reference/plotCophenetic.html">plotCophenetic</a></span>(<span class="kw">res</span> <span class="kw">=</span> <span class="no">laml.sign</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-47-1.png" width="288"></p>
<p>Best possible value is the one at which the correlation value on the y-axis drops significantly. In this case it appears to be at <code>n = 3</code>. LAML is not an ideal example for signature analysis with its low mutation rate, but for solid tumors with higher mutation burden one could expect more signatures, provided sufficient number of samples.</p>
<p>Once <code>n</code> is estimated, we can run the main function.</p>
<div class="sourceCode" id="cb120"><html><body><pre class="r"><span class="no">laml.sig</span> <span class="kw">=</span> <span class="fu"><a href="../reference/extractSignatures.html">extractSignatures</a></span>(<span class="kw">mat</span> <span class="kw">=</span> <span class="no">laml.tnm</span>, <span class="kw">n</span> <span class="kw">=</span> <span class="fl">3</span>)</pre></body></html></div>
<pre><code>## -Running NMF for factorization rank: 3</code></pre>
<pre><code>## -Finished in2.528s elapsed (2.300s cpu)</code></pre>
<p>Compare detected signatures to COSMIC <a href="https://cancer.sanger.ac.uk/cosmic/signatures_v2">Legacy</a> or <a href="https://cancer.sanger.ac.uk/cosmic/signatures_v2">SBS</a> signature database.</p>
<div class="sourceCode" id="cb123"><html><body><pre class="r"><span class="co">#Compate against original 30 signatures </span>
<span class="no">laml.og30.cosm</span> <span class="kw">=</span> <span class="fu"><a href="../reference/compareSignatures.html">compareSignatures</a></span>(<span class="kw">nmfRes</span> <span class="kw">=</span> <span class="no">laml.sig</span>, <span class="kw">sig_db</span> <span class="kw">=</span> <span class="st">"legacy"</span>)</pre></body></html></div>
<pre><code>## -Comparing against COSMIC signatures</code></pre>
<pre><code>## ------------------------------------</code></pre>
<pre><code>## --Found Signature_1 most similar to COSMIC_1</code></pre>
<pre><code>##    Aetiology: spontaneous deamination of 5-methylcytosine [cosine-similarity: 0.84]</code></pre>
<pre><code>## --Found Signature_2 most similar to COSMIC_1</code></pre>
<pre><code>##    Aetiology: spontaneous deamination of 5-methylcytosine [cosine-similarity: 0.577]</code></pre>
<pre><code>## --Found Signature_3 most similar to COSMIC_5</code></pre>
<pre><code>##    Aetiology: Unknown [cosine-similarity: 0.851]</code></pre>
<pre><code>## ------------------------------------</code></pre>
<div class="sourceCode" id="cb133"><html><body><pre class="r"><span class="co">#Compate against updated version3 60 signatures </span>
<span class="no">laml.v3.cosm</span> <span class="kw">=</span> <span class="fu"><a href="../reference/compareSignatures.html">compareSignatures</a></span>(<span class="kw">nmfRes</span> <span class="kw">=</span> <span class="no">laml.sig</span>, <span class="kw">sig_db</span> <span class="kw">=</span> <span class="st">"SBS"</span>)</pre></body></html></div>
<pre><code>## -Comparing against COSMIC signatures
## ------------------------------------</code></pre>
<pre><code>## --Found Signature_1 most similar to SBS1</code></pre>
<pre><code>##    Aetiology: spontaneous or enzymatic deamination of 5-methylcytosine [cosine-similarity: 0.858]</code></pre>
<pre><code>## --Found Signature_2 most similar to SBS6</code></pre>
<pre><code>##    Aetiology: defective DNA mismatch repair [cosine-similarity: 0.538]</code></pre>
<pre><code>## --Found Signature_3 most similar to SBS3</code></pre>
<pre><code>##    Aetiology: Defects in DNA-DSB repair by HR [cosine-similarity: 0.836]</code></pre>
<pre><code>## ------------------------------------</code></pre>
<p><code>compareSignatures</code> returns full table of cosine similarities against COSMIC signatures, which can be further analysed. Below plot shows comparison of similarities of detected signatures against validated signatures.</p>
<div class="sourceCode" id="cb142"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">'pheatmap'</span>)
<span class="kw pkg">pheatmap</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span>(<span class="kw">mat</span> <span class="kw">=</span> <span class="no">laml.og30.cosm</span>$<span class="no">cosine_similarities</span>, <span class="kw">cluster_rows</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">main</span> <span class="kw">=</span> <span class="st">"cosine similarity against validated signatures"</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-51-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Finally plot signatures</p>
<div class="sourceCode" id="cb143"><html><body><pre class="r"><span class="kw pkg">maftools</span><span class="kw ns">::</span><span class="fu"><a href="../reference/plotSignatures.html">plotSignatures</a></span>(<span class="kw">nmfRes</span> <span class="kw">=</span> <span class="no">laml.sig</span>, <span class="kw">title_size</span> <span class="kw">=</span> <span class="fl">0.8</span>, <span class="kw">sig_db</span> <span class="kw">=</span> <span class="st">"SBS"</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-52-1.png" width="576" style="display: block; margin: auto;"></p>
<p>If you fancy 3D barpots, you can install <code>barplot3d</code> package and visualize the results with <code>legoplot3d</code> function.</p>
<div class="sourceCode" id="cb144"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"barplot3d"</span>)
<span class="co">#Visualize first signature</span>
<span class="no">sig1</span> <span class="kw">=</span> <span class="no">laml.sig</span>$<span class="no">signatures</span>[,<span class="fl">1</span>]
<span class="kw pkg">barplot3d</span><span class="kw ns">::</span><span class="fu">legoplot3d</span>(<span class="kw">contextdata</span> <span class="kw">=</span> <span class="no">sig1</span>, <span class="kw">labels</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">scalexy</span> <span class="kw">=</span> <span class="fl">0.01</span>, <span class="kw">sixcolors</span> <span class="kw">=</span> <span class="st">"sanger"</span>, <span class="kw">alpha</span> <span class="kw">=</span> <span class="fl">0.5</span>)</pre></body></html></div>
<p><strong><em>NOTE:</em></strong></p>
<ol style="list-style-type: decimal">
<li><p>Should you receive an error while running <code>extractSignatures</code> complaining <code>none of the packages are loaded</code>, please manually load the <code>NMF</code> library and re-run.</p></li>
<li><p>If either <code>extractSignatures</code> or <code>estimateSignatures</code> stops in between, its possibly due to low mutation counts in the matrix. In that case rerun the functions with <code>pConstant</code> argument set to small positive value (e.g, 0.1).</p></li>
</ol>
</div>
<div id="signature-enrichment-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#signature-enrichment-analysis" class="anchor"></a>Signature enrichment analysis</h3>
<p>Signatures can further be assigned to samples and enrichment analysis can be performd using <code>signatureEnrichment</code> funtion, which identifies mutations enriched in every signature identified.</p>
<div class="sourceCode" id="cb145"><html><body><pre class="r"><span class="no">laml.se</span> <span class="kw">=</span> <span class="fu"><a href="../reference/signatureEnrichment.html">signatureEnrichment</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">sig_res</span> <span class="kw">=</span> <span class="no">laml.sig</span>)</pre></body></html></div>
<pre><code>## Running k-means for signature assignment..</code></pre>
<pre><code>## Performing pairwise and groupwise comparisions..</code></pre>
<pre><code>## Sample size per factor in Signature:</code></pre>
<pre><code>## 
## Signature_1 Signature_2 Signature_3 
##          60          65          63</code></pre>
<pre><code>## Estimating mutation load and signature exposures..</code></pre>
<p><img src="maftools_files/figure-html/unnamed-chunk-55-1.png" width="480"></p>
<p>Above results can be visualzied similar to clinical enrichments.</p>
<div class="sourceCode" id="cb151"><html><body><pre class="r"><span class="fu"><a href="../reference/plotEnrichmentResults.html">plotEnrichmentResults</a></span>(<span class="kw">enrich_res</span> <span class="kw">=</span> <span class="no">laml.se</span>, <span class="kw">pVal</span> <span class="kw">=</span> <span class="fl">0.05</span>)</pre></body></html></div>
<p><img src="maftools_files/figure-html/unnamed-chunk-56-1.png" width="576"></p>
</div>
</div>
</div>
<div id="variant-annotations" class="section level1">
<h1 class="hasAnchor">
<a href="#variant-annotations" class="anchor"></a>Variant Annotations</h1>
<div id="converting-annovar-output-to-maf" class="section level2">
<h2 class="hasAnchor">
<a href="#converting-annovar-output-to-maf" class="anchor"></a>Converting annovar output to MAF</h2>
<p>Annovar is one of the most widely used Variant Annotation tool in Genomics <a href="#references">17</a>. Annovar output is generally in a tabular format with various annotation columns. This function converts such annovar output files into MAF. This function requires that annovar was run with gene based annotation as a first operation, before including any filter or region based annotations.</p>
<p>e.g,</p>
<div class="sourceCode" id="cb152"><pre class="sourceCode perl"><code class="sourceCode perl"><span id="cb152-1"><a href="#cb152-1"></a>table_annovar.pl example/ex1.avinput humandb/ -buildver hg19 -out myanno -remove -protocol (refGene),cytoBand,dbnsfp30a -operation (g),r,f -nastring NA</span></code></pre></div>
<p><code>annovarToMaf</code> mainly uses gene based annotations for processing, rest of the annotation columns from input file will be attached to the end of the resulting MAF.</p>
<p>As an example we will annotate the same file which was used above to run <code>oncotate</code> function. We will annotate it using annovar with the following command. For simplicity, here we are including only gene based annotations but one can include as many annotations as they wish. But make sure the fist operation is always gene based annotation.</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb153-1"><a href="#cb153-1"></a><span class="va">$perl</span> <span class="ex">table_annovar.pl</span> variants.tsv ~/path/to/humandb/ -buildver hg19 </span>
<span id="cb153-2"><a href="#cb153-2"></a><span class="ex">-out</span> variants --otherinfo -remove -protocol ensGene -operation g -nastring NA</span></code></pre></div>
<p>Output generated is stored as a part of this package. We can convert this annovar output into MAF using <code>annovarToMaf</code>.</p>
<div class="sourceCode" id="cb154"><html><body><pre class="r"><span class="no">var.annovar</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"variants.hg19_multianno.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="no">var.annovar.maf</span> <span class="kw">=</span> <span class="fu"><a href="../reference/annovarToMaf.html">annovarToMaf</a></span>(<span class="kw">annovar</span> <span class="kw">=</span> <span class="no">var.annovar</span>, <span class="kw">Center</span> <span class="kw">=</span> <span class="st">'CSI-NUS'</span>, <span class="kw">refBuild</span> <span class="kw">=</span> <span class="st">'hg19'</span>,
                               <span class="kw">tsbCol</span> <span class="kw">=</span> <span class="st">'Tumor_Sample_Barcode'</span>, <span class="kw">table</span> <span class="kw">=</span> <span class="st">'ensGene'</span>)</pre></body></html></div>
<pre><code>## -Reading annovar files
## -Processing Exonic variants
## --Adding Variant_Classification
## --Parsing aa-change
## -Processing Non-exonic variants
## --Adding Variant_Classification
## -Adding Variant_Type
## -Converting Ensemble Gene IDs into HGNC gene symbols
## --Done. Original ensemble gene IDs are preserved under field name ens_id
## Finished in 0.128s elapsed (0.115s cpu)</code></pre>
<p>Annovar, when used with Ensemble as a gene annotation source, uses ensemble gene IDs as Gene names. In that case, use <code>annovarToMaf</code> with argument <code>table</code> set to <code>ensGene</code> which converts ensemble gene IDs into HGNC symbols.</p>
</div>
<div id="converting-icgc-simple-somatic-mutation-format-to-maf" class="section level2">
<h2 class="hasAnchor">
<a href="#converting-icgc-simple-somatic-mutation-format-to-maf" class="anchor"></a>Converting ICGC Simple Somatic Mutation Format to MAF</h2>
<p>Just like TCGA, International Cancer Genome Consortium <a href="http://icgc.org">ICGC</a> also makes its data publicly available. But the data are stored in <a href="http://docs.icgc.org/submission/guide/icgc-simple-somatic-mutation-format/">Simpale Somatic Mutation Format</a>, which is similar to MAF format in its structure. However field names and classification of variants is different from that of MAF. <code>icgcSimpleMutationToMAF</code> is a function which reads ICGC data and converts them to MAF.</p>
<div class="sourceCode" id="cb156"><html><body><pre class="r"><span class="co">#Read sample ICGC data for ESCA</span>
<span class="no">esca.icgc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"simple_somatic_mutation.open.ESCA-CN.sample.tsv.gz"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="no">esca.maf</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/icgcSimpleMutationToMAF.html">icgcSimpleMutationToMAF</a></span>(<span class="kw">icgc</span> <span class="kw">=</span> <span class="no">esca.icgc</span>, <span class="kw">addHugoSymbol</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<pre><code>## Converting Ensemble Gene IDs into HGNC gene symbols.</code></pre>
<pre><code>## Done! Original ensemble gene IDs are preserved under field name ens_id</code></pre>
<pre><code>## --Removed 427 duplicated variants</code></pre>
<div class="sourceCode" id="cb160"><html><body><pre class="r"><span class="co">#Printing first 16 columns for display convenience.</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">esca.maf</span>[<span class="fl">1</span>:<span class="fl">5</span>,<span class="fl">1</span>:<span class="fl">16</span>, <span class="kw">with</span> <span class="kw">=</span> <span class="fl">FALSE</span>])</pre></body></html></div>
<pre><code>##    Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome Start_Position
## 1:  AC005237.4             NA     NA     GRCh37          2      241987787
## 2:  AC061992.1            786     NA     GRCh37         17       76425382
## 3:  AC097467.2             NA     NA     GRCh37          4      156294567
## 4:    ADAMTS12             NA     NA     GRCh37          5       33684019
## 5:  AL589642.1          54801     NA     GRCh37          9       32630154
##    End_Position Strand Variant_Classification Variant_Type Reference_Allele
## 1:    241987787      +                 Intron          SNP                C
## 2:     76425382      +                3'Flank          SNP                C
## 3:    156294567      +                 Intron          SNP                A
## 4:     33684019      +      Missense_Mutation          SNP                A
## 5:     32630154      +                    RNA          SNP                T
##    Tumor_Seq_Allele1 Tumor_Seq_Allele2 dbSNP_RS dbSNP_Val_Status
## 1:                 C                 T       NA               NA
## 2:                 C                 T       NA               NA
## 3:                 A                 G       NA               NA
## 4:                 A                 C       NA               NA
## 5:                 T                 C       NA               NA
##    Tumor_Sample_Barcode
## 1:             SA514619
## 2:             SA514619
## 3:             SA514619
## 4:             SA514619
## 5:             SA514619</code></pre>
<p>Note that by default Simple Somatic Mutation format contains all affected transcripts of a variant resulting in multiple entries of the same variant in same sample. It is hard to choose a single affected transcript based on annotations alone and by default this program removes repeated variants as duplicated entries. If you wish to keep all of them, set <code>removeDuplicatedVariants</code> to FALSE. Another option is to convert input file to MAF by removing duplicates and then use scripts like <a href="https://github.com/mskcc/vcf2maf">vcf2maf</a> to re-annotate and prioritize affected transcripts.</p>
</div>
<div id="prepare-maf-file-for-mutsigcv-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#prepare-maf-file-for-mutsigcv-analysis" class="anchor"></a>Prepare MAF file for MutSigCV analysis</h2>
<p>MutSig/MutSigCV is most widely used program for detecting driver genes <a href="#references">18</a>. However, we have observed that covariates files (gene.covariates.txt and exome_full192.coverage.txt) which are bundled with MutSig have non-standard gene names (non Hugo_Symbols). This discrepancy between Hugo_Symbols in MAF and non-Hugo_symbols in covariates file causes MutSig program to ignore such genes. For example, KMT2D - a well known driver gene in Esophageal Carcinoma is represented as MLL2 in MutSig covariates. This causes KMT2D to be ignored from analysis and is represented as an insignificant gene in MutSig results. This function attempts to correct such gene symbols with a manually curated list of gene names compatible with MutSig covariates list.</p>
<div class="sourceCode" id="cb162"><html><body><pre class="r"><span class="no">laml.mutsig.corrected</span> <span class="kw">=</span> <span class="fu"><a href="../reference/prepareMutSig.html">prepareMutSig</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>)
<span class="co"># Converting gene names for 1 variants from 1 genes</span>
<span class="co">#    Hugo_Symbol MutSig_Synonym N</span>
<span class="co"># 1:    ARHGAP35          GRLF1 1</span>
<span class="co"># Original symbols are preserved under column OG_Hugo_Symbol.</span></pre></body></html></div>
</div>
</div>
<div id="set-operations" class="section level1">
<h1 class="hasAnchor">
<a href="#set-operations" class="anchor"></a>Set operations</h1>
<div id="subsetting-maf" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-maf" class="anchor"></a>Subsetting MAF</h2>
<p>We can also subset MAF using function <code>subsetMaf</code></p>
<div class="sourceCode" id="cb163"><html><body><pre class="r"><span class="co">#Extract data for samples 'TCGA.AB.3009' and 'TCGA.AB.2933'  (Printing just 5 rows for display convenience)</span>
<span class="fu"><a href="../reference/subsetMaf.html">subsetMaf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">tsb</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'TCGA-AB-3009'</span>, <span class="st">'TCGA-AB-2933'</span>), <span class="kw">mafObj</span> <span class="kw">=</span> <span class="fl">FALSE</span>)[<span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##    Hugo_Symbol Entrez_Gene_Id           Center NCBI_Build Chromosome
## 1:      ABCB11           8647 genome.wustl.edu         37          2
## 2:       ACSS3          79611 genome.wustl.edu         37         12
## 3:        ANK3            288 genome.wustl.edu         37         10
## 4:       AP1G2           8906 genome.wustl.edu         37         14
## 5:         ARC          23237 genome.wustl.edu         37          8
##    Start_Position End_Position Strand Variant_Classification Variant_Type
## 1:      169780250    169780250      +      Missense_Mutation          SNP
## 2:       81536902     81536902      +      Missense_Mutation          SNP
## 3:       61926581     61926581      +            Splice_Site          SNP
## 4:       24033309     24033309      +      Missense_Mutation          SNP
## 5:      143694874    143694874      +      Missense_Mutation          SNP
##    Reference_Allele Tumor_Seq_Allele1 Tumor_Seq_Allele2 Tumor_Sample_Barcode
## 1:                G                 G                 A         TCGA-AB-3009
## 2:                C                 C                 T         TCGA-AB-3009
## 3:                C                 C                 A         TCGA-AB-3009
## 4:                C                 C                 T         TCGA-AB-3009
## 5:                C                 C                 G         TCGA-AB-3009
##    Protein_Change i_TumorVAF_WU i_transcript_name
## 1:       p.A1283V      46.97218       NM_003742.2
## 2:        p.A266V      47.32510       NM_024560.2
## 3:                     43.99478       NM_020987.2
## 4:        p.R346Q      47.08000       NM_003917.2
## 5:        p.W253C      42.96435       NM_015193.3</code></pre>
<div class="sourceCode" id="cb165"><html><body><pre class="r"><span class="co">##Same as above but return output as an MAF object (Default behaviour)</span>
<span class="fu"><a href="../reference/subsetMaf.html">subsetMaf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">tsb</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'TCGA-AB-3009'</span>, <span class="st">'TCGA-AB-2933'</span>))</pre></body></html></div>
<pre><code>## An object of class  MAF 
##                    ID          summary Mean Median
##  1:        NCBI_Build               37   NA     NA
##  2:            Center genome.wustl.edu   NA     NA
##  3:           Samples                2   NA     NA
##  4:            nGenes               34   NA     NA
##  5:   Frame_Shift_Ins                5  2.5    2.5
##  6:      In_Frame_Ins                1  0.5    0.5
##  7: Missense_Mutation               26 13.0   13.0
##  8: Nonsense_Mutation                2  1.0    1.0
##  9:       Splice_Site                1  0.5    0.5
## 10:             total               35 17.5   17.5</code></pre>
<div id="specifying-queries-and-controlling-output-fields-" class="section level3">
<h3 class="hasAnchor">
<a href="#specifying-queries-and-controlling-output-fields-" class="anchor"></a>Specifying queries and controlling output fields.</h3>
<div class="sourceCode" id="cb167"><html><body><pre class="r"><span class="co">#Select all Splice_Site mutations from DNMT3A and NPM1</span>
<span class="fu"><a href="../reference/subsetMaf.html">subsetMaf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">genes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'DNMT3A'</span>, <span class="st">'NPM1'</span>), <span class="kw">mafObj</span> <span class="kw">=</span> <span class="fl">FALSE</span>,<span class="kw">query</span> <span class="kw">=</span> <span class="st">"Variant_Classification == 'Splice_Site'"</span>)</pre></body></html></div>
<pre><code>##    Hugo_Symbol Entrez_Gene_Id           Center NCBI_Build Chromosome
## 1:      DNMT3A           1788 genome.wustl.edu         37          2
## 2:      DNMT3A           1788 genome.wustl.edu         37          2
## 3:      DNMT3A           1788 genome.wustl.edu         37          2
## 4:      DNMT3A           1788 genome.wustl.edu         37          2
## 5:      DNMT3A           1788 genome.wustl.edu         37          2
## 6:      DNMT3A           1788 genome.wustl.edu         37          2
##    Start_Position End_Position Strand Variant_Classification Variant_Type
## 1:       25459874     25459874      +            Splice_Site          SNP
## 2:       25467208     25467208      +            Splice_Site          SNP
## 3:       25467022     25467022      +            Splice_Site          SNP
## 4:       25459803     25459803      +            Splice_Site          SNP
## 5:       25464576     25464576      +            Splice_Site          SNP
## 6:       25469029     25469029      +            Splice_Site          SNP
##    Reference_Allele Tumor_Seq_Allele1 Tumor_Seq_Allele2 Tumor_Sample_Barcode
## 1:                C                 C                 G         TCGA-AB-2818
## 2:                C                 C                 T         TCGA-AB-2822
## 3:                A                 A                 G         TCGA-AB-2891
## 4:                A                 A                 C         TCGA-AB-2898
## 5:                C                 C                 A         TCGA-AB-2934
## 6:                C                 C                 A         TCGA-AB-2968
##    Protein_Change i_TumorVAF_WU i_transcript_name
## 1:        p.R803S         36.84       NM_022552.3
## 2:                        62.96       NM_022552.3
## 3:                        34.78       NM_022552.3
## 4:                        46.43       NM_022552.3
## 5:        p.G646V         37.50       NM_022552.3
## 6:        p.E477*         40.00       NM_022552.3</code></pre>
<div class="sourceCode" id="cb169"><html><body><pre class="r"><span class="co">#Same as above but include only 'i_transcript_name' column in the output</span>
<span class="fu"><a href="../reference/subsetMaf.html">subsetMaf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">genes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'DNMT3A'</span>, <span class="st">'NPM1'</span>), <span class="kw">mafObj</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">query</span> <span class="kw">=</span> <span class="st">"Variant_Classification == 'Splice_Site'"</span>, <span class="kw">fields</span> <span class="kw">=</span> <span class="st">'i_transcript_name'</span>)</pre></body></html></div>
<pre><code>##    Hugo_Symbol Chromosome Start_Position End_Position Reference_Allele
## 1:      DNMT3A          2       25459874     25459874                C
## 2:      DNMT3A          2       25467208     25467208                C
## 3:      DNMT3A          2       25467022     25467022                A
## 4:      DNMT3A          2       25459803     25459803                A
## 5:      DNMT3A          2       25464576     25464576                C
## 6:      DNMT3A          2       25469029     25469029                C
##    Tumor_Seq_Allele2 Variant_Classification Variant_Type Tumor_Sample_Barcode
## 1:                 G            Splice_Site          SNP         TCGA-AB-2818
## 2:                 T            Splice_Site          SNP         TCGA-AB-2822
## 3:                 G            Splice_Site          SNP         TCGA-AB-2891
## 4:                 C            Splice_Site          SNP         TCGA-AB-2898
## 5:                 A            Splice_Site          SNP         TCGA-AB-2934
## 6:                 A            Splice_Site          SNP         TCGA-AB-2968
##    i_transcript_name
## 1:       NM_022552.3
## 2:       NM_022552.3
## 3:       NM_022552.3
## 4:       NM_022552.3
## 5:       NM_022552.3
## 6:       NM_022552.3</code></pre>
</div>
<div id="subsetting-by-clinical-data" class="section level3">
<h3 class="hasAnchor">
<a href="#subsetting-by-clinical-data" class="anchor"></a>Subsetting by clinical data</h3>
<p>Use <code>clinQuery</code> argument in <code>subsetMaf</code> to select samples of interest based on their clinical features.</p>
<div class="sourceCode" id="cb171"><html><body><pre class="r"><span class="co">#Select all samples with FAB clasification M4 in clinical data </span>
<span class="no">laml_m4</span> <span class="kw">=</span> <span class="fu"><a href="../reference/subsetMaf.html">subsetMaf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">clinQuery</span> <span class="kw">=</span> <span class="st">"FAB_classification %in% 'M4'"</span>)</pre></body></html></div>
<pre><code>## -subsetting by clinical data..</code></pre>
<pre><code>## --39 samples meet the clinical query</code></pre>
</div>
</div>
</div>
<div id="pre-compiled-tcga-maf-objects" class="section level1">
<h1 class="hasAnchor">
<a href="#pre-compiled-tcga-maf-objects" class="anchor"></a>Pre-compiled TCGA MAF objects</h1>
<p>There is also an R data package containing pre-compiled TCGA MAF objects from TCGA firehose and TCGA <a href="https://gdc.cancer.gov/about-data/publications/mc3-2017">MC3</a> projects, particularly helpful for those working with TCGA mutation data. Every dataset is stored as an MAF object containing somatic mutations along with clinical information. Due to Bioconductor package size limits and other difficulties, this was not submitted to Bioconductor. However, you can still download <a href="https://github.com/PoisonAlien/TCGAmutations">TCGAmutations</a> package from GitHub.</p>
<div class="sourceCode" id="cb174"><html><body><pre class="r"><span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="kw">repo</span> <span class="kw">=</span> <span class="st">"PoisonAlien/TCGAmutations"</span>)</pre></body></html></div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<ol style="list-style-type: decimal">
<li>Cancer Genome Atlas Research, N. Genomic and epigenomic landscapes of adult de novo acute myeloid leukemia. N Engl J Med 368, 2059-74 (2013).</li>
<li>Mermel, C.H. et al. GISTIC2.0 facilitates sensitive and confident localization of the targets of focal somatic copy-number alteration in human cancers. Genome Biol 12, R41 (2011).</li>
<li>Olshen, A.B., Venkatraman, E.S., Lucito, R. &amp; Wigler, M. Circular binary segmentation for the analysis of array-based DNA copy number data. Biostatistics 5, 557-72 (2004).</li>
<li>Alexandrov, L.B. et al. Signatures of mutational processes in human cancer. Nature 500, 415-21 (2013).</li>
<li>Tamborero, D., Gonzalez-Perez, A. &amp; Lopez-Bigas, N. OncodriveCLUST: exploiting the positional clustering of somatic mutations to identify cancer genes. Bioinformatics 29, 2238-44 (2013).</li>
<li>Dees, N.D. et al. MuSiC: identifying mutational significance in cancer genomes. Genome Res 22, 1589-98 (2012).</li>
<li>Lawrence MS, Stojanov P, Mermel CH, Robinson JT, Garraway LA, Golub TR, Meyerson M, Gabriel SB, Lander ES, Getz G: Discovery and saturation analysis of cancer genes across 21 tumour types. Nature 2014, 505:495-501.</li>
<li>Griffith, M., Griffith, O. L., Coffman, A. C., Weible, J. V., McMichael, J. F., Spies, N. C., … Wilson, R. K. (2013). DGIdb - Mining the druggable genome. Nature Methods, 10(12), 1209–1210. <a href="http://doi.org/10.1038/nmeth.2689" class="uri">http://doi.org/10.1038/nmeth.2689</a>
</li>
<li>Sanchez-Vega F, Mina M, Armenia J, Chatila WK, Luna A, La KC, Dimitriadoy S, Liu DL, Kantheti HS, Saghafinia S et al. 2018. Oncogenic Signaling Pathways in The Cancer Genome Atlas. Cell 173: 321-337 e310</li>
<li>Madan, V. et al. Comprehensive mutational analysis of primary and relapse acute promyelocytic leukemia. Leukemia 30, 1672-81 (2016).</li>
<li>Mroz, E.A. &amp; Rocco, J.W. MATH, a novel measure of intratumor genetic heterogeneity, is high in poor-outcome classes of head and neck squamous cell carcinoma. Oral Oncol 49, 211-5 (2013).</li>
<li>Roberts SA, Lawrence MS, Klimczak LJ, et al. An APOBEC Cytidine Deaminase Mutagenesis Pattern is Widespread in Human Cancers. Nature genetics. 2013;45(9):970-976.</li>
<li>Gaujoux, R. &amp; Seoighe, C. A flexible R package for nonnegative matrix factorization. BMC Bioinformatics 11, 367 (2010).</li>
<li>Welch, J.S. et al. The origin and evolution of mutations in acute myeloid leukemia. Cell 150, 264-78 (2012).</li>
<li>Ramos, A.H. et al. Oncotator: cancer variant annotation tool. Hum Mutat 36, E2423-9 (2015).</li>
<li>Wang, K., Li, M. &amp; Hakonarson, H. ANNOVAR: functional annotation of genetic variants from high-throughput sequencing data. Nucleic Acids Res 38, e164 (2010).</li>
<li>Lawrence MS, Stojanov P, Polak P, Kryukov GV, Cibulskis K, Sivachenko A, Carter SL, Stewart C, Mermel CH, Roberts SA, et al: Mutational heterogeneity in cancer and the search for new cancer-associated genes. Nature 2013, 499:214-218.</li>
</ol>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb175"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R version 4.0.2 (2020-06-22)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Catalina 10.15.6
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] pheatmap_1.0.12                   doParallel_1.0.15                
##  [3] iterators_1.0.12                  foreach_1.5.0                    
##  [5] NMF_0.23.0                        Biobase_2.48.0                   
##  [7] cluster_2.1.0                     rngtools_1.5                     
##  [9] pkgmaker_0.31.1                   registry_0.5-1                   
## [11] BSgenome.Hsapiens.UCSC.hg19_1.4.3 BSgenome_1.56.0                  
## [13] rtracklayer_1.48.0                Biostrings_2.56.0                
## [15] XVector_0.28.0                    GenomicRanges_1.40.0             
## [17] GenomeInfoDb_1.24.2               IRanges_2.22.2                   
## [19] S4Vectors_0.26.1                  BiocGenerics_0.34.0              
## [21] mclust_5.4.6                      maftools_2.4.10                  
## 
## loaded via a namespace (and not attached):
##  [1] splines_4.0.2               R.utils_2.9.2              
##  [3] assertthat_0.2.1            GenomeInfoDbData_1.2.3     
##  [5] Rsamtools_2.4.0             yaml_2.2.1                 
##  [7] pillar_1.4.6                backports_1.1.8            
##  [9] lattice_0.20-41             glue_1.4.1                 
## [11] digest_0.6.25               RColorBrewer_1.1-2         
## [13] colorspace_1.4-1            plyr_1.8.6                 
## [15] htmltools_0.5.0             Matrix_1.2-18              
## [17] R.oo_1.23.0                 pkgconfig_2.0.3            
## [19] XML_3.99-0.5                bibtex_0.4.2.2             
## [21] zlibbioc_1.34.0             purrr_0.3.4                
## [23] xtable_1.8-4                scales_1.1.1               
## [25] BiocParallel_1.22.0         tibble_3.0.3               
## [27] generics_0.0.2              ggplot2_3.3.2              
## [29] ellipsis_0.3.1              withr_2.2.0                
## [31] SummarizedExperiment_1.18.2 survival_3.2-3             
## [33] magrittr_1.5                crayon_1.3.4               
## [35] memoise_1.1.0               evaluate_0.14              
## [37] R.methodsS3_1.8.0           fs_1.5.0                   
## [39] MASS_7.3-51.6               tools_4.0.2                
## [41] data.table_1.13.0           gridBase_0.4-7             
## [43] lifecycle_0.2.0             matrixStats_0.56.0         
## [45] stringr_1.4.0               munsell_0.5.0              
## [47] DelayedArray_0.14.1         compiler_4.0.2             
## [49] pkgdown_1.5.1               rlang_0.4.7                
## [51] grid_4.0.2                  RCurl_1.98-1.2             
## [53] rstudioapi_0.11             bitops_1.0-6               
## [55] rmarkdown_2.3               gtable_0.3.0               
## [57] codetools_0.2-16            reshape2_1.4.4             
## [59] R6_2.4.1                    GenomicAlignments_1.24.0   
## [61] dplyr_1.0.1                 knitr_1.29                 
## [63] rprojroot_1.3-2             desc_1.2.0                 
## [65] stringi_1.4.6               Rcpp_1.0.5                 
## [67] vctrs_0.3.2                 tidyselect_1.1.0           
## [69] xfun_0.16</code></pre>
</div>
<div id="support-and-acknowledgements" class="section level1">
<h1 class="hasAnchor">
<a href="#support-and-acknowledgements" class="anchor"></a>Support and acknowledgements</h1>
<div id="github" class="section level2">
<h2 class="hasAnchor">
<a href="#github" class="anchor"></a>Github</h2>
<p>If you have any issues, bug reports or feature requests please feel free to raise an <a href="https://github.com/PoisonAlien/maftools/issues">issue</a> on <a href="https://github.com/PoisonAlien/maftools">GitHub</a> page.</p>
</div>
<div id="acknowledgements" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h2>
<ul>
<li>Thanks to <a href="https://github.com/ShixiangWang">Shixiang Wang</a> for many bug fixes, improvements, feature requests, and contribution to overall package maintenance</li>
<li>Thanks to <a href="https://github.com/rdmorin">Ryan Morin</a> for crucial bug fixes</li>
<li>Thanks to <a href="https://twitter.com/ellis2013nz">Peter Ellis</a> for beautiful <a href="http://ellisp.github.io/blog/2017/09/09/rmarkdown">markdown template</a>
</li>
<li>
<code>somaticInteractions</code> plot is inspired from the article <a href="https://www.nature.com/articles/ncomms6901">Combining gene mutation with gene expression data improves outcome prediction in myelodysplastic syndromes</a>. Thanks to authors for making source code available.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Anand Mayakonda.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
