<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Customizing oncoplots â€¢ maftools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Customizing oncoplots">
<meta property="og:description" content="maftools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maftools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.4.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/maftools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/oncoplots.html">Customizing oncoplots</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PoisonAlien/maftools/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="oncoplots_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Customizing oncoplots</h1>
                        <h4 class="author">Anand Mayakonda</h4>
            
            <h4 class="date">2020-08-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/PoisonAlien/maftools/blob/master/vignettes/oncoplots.Rmd"><code>vignettes/oncoplots.Rmd</code></a></small>
      <div class="hidden name"><code>oncoplots.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">maftools</span>)</pre></body></html></div>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co">#path to TCGA LAML MAF file</span>
<span class="no">laml.maf</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'tcga_laml.maf.gz'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'maftools'</span>)
<span class="co">#clinical information containing survival information and histology. This is optional</span>
<span class="no">laml.clin</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'tcga_laml_annot.tsv'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'maftools'</span>)

<span class="no">laml</span> <span class="kw">=</span> <span class="fu"><a href="../reference/read.maf.html">read.maf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml.maf</span>,
                <span class="kw">clinicalData</span> <span class="kw">=</span> <span class="no">laml.clin</span>,
                <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<div id="including-transitiontransversions-into-oncoplot" class="section level2">
<h2 class="hasAnchor">
<a href="#including-transitiontransversions-into-oncoplot" class="anchor"></a>Including Transition/Transversions into oncoplot</h2>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="co">#By default the function plots top20 mutated genes</span>
<span class="fu"><a href="../reference/oncoplot.html">oncoplot</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">draw_titv</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="oncoplots_files/figure-html/unnamed-chunk-3-1.png" width="960" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="changing-colors-for-variant-classifications" class="section level2">
<h2 class="hasAnchor">
<a href="#changing-colors-for-variant-classifications" class="anchor"></a>Changing colors for variant classifications</h2>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co">#One can use any colors, here in this example color palette from RColorBrewer package is used</span>
<span class="no">vc_cols</span> <span class="kw">=</span> <span class="kw pkg">RColorBrewer</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span>(<span class="kw">n</span> <span class="kw">=</span> <span class="fl">8</span>, <span class="kw">name</span> <span class="kw">=</span> <span class="st">'Paired'</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">vc_cols</span>) <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(
  <span class="st">'Frame_Shift_Del'</span>,
  <span class="st">'Missense_Mutation'</span>,
  <span class="st">'Nonsense_Mutation'</span>,
  <span class="st">'Multi_Hit'</span>,
  <span class="st">'Frame_Shift_Ins'</span>,
  <span class="st">'In_Frame_Ins'</span>,
  <span class="st">'Splice_Site'</span>,
  <span class="st">'In_Frame_Del'</span>
)

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">vc_cols</span>)
<span class="co">#&gt;   Frame_Shift_Del Missense_Mutation Nonsense_Mutation         Multi_Hit </span>
<span class="co">#&gt;         "#A6CEE3"         "#1F78B4"         "#B2DF8A"         "#33A02C" </span>
<span class="co">#&gt;   Frame_Shift_Ins      In_Frame_Ins       Splice_Site      In_Frame_Del </span>
<span class="co">#&gt;         "#FB9A99"         "#E31A1C"         "#FDBF6F"         "#FF7F00"</span>

<span class="fu"><a href="../reference/oncoplot.html">oncoplot</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">colors</span> <span class="kw">=</span> <span class="no">vc_cols</span>, <span class="kw">top</span> <span class="kw">=</span> <span class="fl">10</span>)</pre></body></html></div>
<p><img src="oncoplots_files/figure-html/unnamed-chunk-4-1.png" width="960" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="including-copy-number-data-into-oncoplots-" class="section level2">
<h2 class="hasAnchor">
<a href="#including-copy-number-data-into-oncoplots-" class="anchor"></a>Including copy number data into oncoplots.</h2>
<p>There are two ways one include CN status into MAF. 1. GISTIC <a href="ftp://ftp.broadinstitute.org/pub/GISTIC2.0/GISTICDocumentation_standalone.htm">results</a> 2. Custom copy number table</p>
<div id="gistic-results" class="section level3">
<h3 class="hasAnchor">
<a href="#gistic-results" class="anchor"></a>GISTIC results</h3>
<p>Most widely used tool for copy number analysis from large scale studies is GISTIC and we can simultaneously read gistic results along with MAF. GISTIC generates numerous files but we need mainly four files <code>all_lesions.conf_XX.txt</code>, <code>amp_genes.conf_XX.txt</code>, <code>del_genes.conf_XX.txt</code>, <code>scores.gistic</code> where XX is confidence level. These files contain significantly altered genomic regions along with amplified and deleted genes respectively.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co">#GISTIC results LAML</span>
<span class="no">all.lesions</span> <span class="kw">=</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"all_lesions.conf_99.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="no">amp.genes</span> <span class="kw">=</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"amp_genes.conf_99.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="no">del.genes</span> <span class="kw">=</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"del_genes.conf_99.txt"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="no">scores.gis</span> <span class="kw">=</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"scores.gistic"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)

<span class="co">#Read GISTIC results along with MAF</span>
<span class="no">laml.plus.gistic</span> <span class="kw">=</span> <span class="fu"><a href="../reference/read.maf.html">read.maf</a></span>(
  <span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml.maf</span>,
  <span class="kw">gisticAllLesionsFile</span> <span class="kw">=</span> <span class="no">all.lesions</span>,
  <span class="kw">gisticAmpGenesFile</span> <span class="kw">=</span> <span class="no">amp.genes</span>,
  <span class="kw">gisticDelGenesFile</span> <span class="kw">=</span> <span class="no">del.genes</span>,
  <span class="kw">gisticScoresFile</span> <span class="kw">=</span> <span class="no">scores.gis</span>,
  <span class="kw">isTCGA</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
  <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
  <span class="kw">clinicalData</span> <span class="kw">=</span> <span class="no">laml.clin</span>
)</pre></body></html></div>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="../reference/oncoplot.html">oncoplot</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml.plus.gistic</span>, <span class="kw">top</span> <span class="kw">=</span> <span class="fl">10</span>)</pre></body></html></div>
<p><img src="oncoplots_files/figure-html/unnamed-chunk-6-1.png" width="960" style="display: block; margin: auto auto auto 0;"></p>
<p>This plot shows frequent deletions in TP53 gene which is located on one of the significantly deleted locus 17p13.2.</p>
</div>
<div id="custom-copy-number-table" class="section level3">
<h3 class="hasAnchor">
<a href="#custom-copy-number-table" class="anchor"></a>Custom copy-number table</h3>
<p>In case there is no GISTIC results available, one can generate a table containing CN status for known genes in known samples. This can be easily created and read along with MAF file.</p>
<p>For example lets create a dummy CN alterations for <code>DNMT3A</code> in random 20 samples.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="kw">seed</span> <span class="kw">=</span> <span class="fl">1024</span>)
<span class="no">barcodes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="fu"><a href="../reference/getSampleSummary.html">getSampleSummary</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">laml</span>)[,<span class="no">Tumor_Sample_Barcode</span>])
<span class="co">#Random 20 samples</span>
<span class="no">dummy.samples</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">barcodes</span>,
                       <span class="kw">size</span> <span class="kw">=</span> <span class="fl">20</span>,
                       <span class="kw">replace</span> <span class="kw">=</span> <span class="fl">FALSE</span>)

<span class="co">#Genarate random CN status for above samples</span>
<span class="no">cn.status</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(
  <span class="kw">x</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'ShallowAmp'</span>, <span class="st">'DeepDel'</span>, <span class="st">'Del'</span>, <span class="st">'Amp'</span>),
  <span class="kw">size</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">dummy.samples</span>),
  <span class="kw">replace</span> <span class="kw">=</span> <span class="fl">TRUE</span>
)

<span class="no">custom.cn.data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(
  <span class="kw">Gene</span> <span class="kw">=</span> <span class="st">"DNMT3A"</span>,
  <span class="kw">Sample_name</span> <span class="kw">=</span> <span class="no">dummy.samples</span>,
  <span class="kw">CN</span> <span class="kw">=</span> <span class="no">cn.status</span>,
  <span class="kw">stringsAsFactors</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">custom.cn.data</span>)
<span class="co">#&gt;     Gene  Sample_name         CN</span>
<span class="co">#&gt; 1 DNMT3A TCGA-AB-2898 ShallowAmp</span>
<span class="co">#&gt; 2 DNMT3A TCGA-AB-2879        Del</span>
<span class="co">#&gt; 3 DNMT3A TCGA-AB-2920        Amp</span>
<span class="co">#&gt; 4 DNMT3A TCGA-AB-2866        Del</span>
<span class="co">#&gt; 5 DNMT3A TCGA-AB-2892        Del</span>
<span class="co">#&gt; 6 DNMT3A TCGA-AB-2863 ShallowAmp</span>

<span class="no">laml.plus.cn</span> <span class="kw">=</span> <span class="fu"><a href="../reference/read.maf.html">read.maf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml.maf</span>,
                        <span class="kw">cnTable</span> <span class="kw">=</span> <span class="no">custom.cn.data</span>,
                        <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)

<span class="fu"><a href="../reference/oncoplot.html">oncoplot</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml.plus.cn</span>, <span class="kw">top</span> <span class="kw">=</span> <span class="fl">5</span>)</pre></body></html></div>
<p><img src="oncoplots_files/figure-html/unnamed-chunk-7-1.png" width="960" style="display: block; margin: auto auto auto 0;"></p>
</div>
</div>
<div id="bar-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#bar-plots" class="anchor"></a>Bar plots</h2>
<p><code>leftBarData</code>, <code>rightBarData</code> and <code>topBarData</code> arguments can be used to display additional values as barplots. Below example demonstrates adding gene expression values and mutsig q-values as left and right side bars respectivelly.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="co">#Selected AML driver genes</span>
<span class="no">aml_genes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"TP53"</span>, <span class="st">"WT1"</span>, <span class="st">"PHF6"</span>, <span class="st">"DNMT3A"</span>, <span class="st">"DNMT3B"</span>, <span class="st">"TET1"</span>, <span class="st">"TET2"</span>, <span class="st">"IDH1"</span>, <span class="st">"IDH2"</span>, <span class="st">"FLT3"</span>, <span class="st">"KIT"</span>, <span class="st">"KRAS"</span>, <span class="st">"NRAS"</span>, <span class="st">"RUNX1"</span>, <span class="st">"CEBPA"</span>, <span class="st">"ASXL1"</span>, <span class="st">"EZH2"</span>, <span class="st">"KDM6A"</span>)

<span class="co">#Variant allele frequcnies (Right bar plot)</span>
<span class="no">aml_genes_vaf</span> <span class="kw">=</span> <span class="fu"><a href="../reference/subsetMaf.html">subsetMaf</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">genes</span> <span class="kw">=</span> <span class="no">aml_genes</span>, <span class="kw">fields</span> <span class="kw">=</span> <span class="st">"i_TumorVAF_WU"</span>, <span class="kw">mafObj</span> <span class="kw">=</span> <span class="fl">FALSE</span>)[,<span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="no">i_TumorVAF_WU</span>, <span class="kw">na.rm</span> <span class="kw">=</span> <span class="fl">TRUE</span>), <span class="no">Hugo_Symbol</span>]
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">aml_genes_vaf</span>)[<span class="fl">2</span>] <span class="kw">=</span> <span class="st">"VAF"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">aml_genes_vaf</span>)
<span class="co">#&gt;    Hugo_Symbol      VAF</span>
<span class="co">#&gt; 1:       ASXL1 37.11250</span>
<span class="co">#&gt; 2:       CEBPA 22.00235</span>
<span class="co">#&gt; 3:      DNMT3A 43.51556</span>
<span class="co">#&gt; 4:      DNMT3B 37.14000</span>
<span class="co">#&gt; 5:        EZH2 68.88500</span>
<span class="co">#&gt; 6:        FLT3 34.60294</span>

<span class="co">#MutSig results (Right bar plot)</span>
<span class="no">laml.mutsig</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"LAML_sig_genes.txt.gz"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"maftools"</span>)
<span class="no">laml.mutsig</span> <span class="kw">=</span> <span class="kw pkg">data.table</span><span class="kw ns">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span>(<span class="kw">input</span> <span class="kw">=</span> <span class="no">laml.mutsig</span>)[,<span class="fu">.</span>(<span class="no">gene</span>, <span class="no">q</span>)]
<span class="no">laml.mutsig</span>[,<span class="no">q</span> <span class="kw">:=</span> -<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="no">q</span>)] <span class="co">#transoform to log10</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">laml.mutsig</span>)
<span class="co">#&gt;      gene        q</span>
<span class="co">#&gt; 1:   FLT3 12.64176</span>
<span class="co">#&gt; 2: DNMT3A 12.64176</span>
<span class="co">#&gt; 3:   NPM1 12.64176</span>
<span class="co">#&gt; 4:   IDH2 12.64176</span>
<span class="co">#&gt; 5:   IDH1 12.64176</span>
<span class="co">#&gt; 6:   TET2 12.64176</span>


<span class="fu"><a href="../reference/oncoplot.html">oncoplot</a></span>(
  <span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>,
  <span class="kw">genes</span> <span class="kw">=</span> <span class="no">aml_genes</span>,
  <span class="kw">leftBarData</span> <span class="kw">=</span> <span class="no">aml_genes_vaf</span>,
  <span class="kw">leftBarLims</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">100</span>),
  <span class="kw">rightBarData</span> <span class="kw">=</span> <span class="no">laml.mutsig</span>,
  <span class="kw">rightBarLims</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">20</span>)
)</pre></body></html></div>
<p><img src="oncoplots_files/figure-html/unnamed-chunk-8-1.png" width="960" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="including-annotations" class="section level2">
<h2 class="hasAnchor">
<a href="#including-annotations" class="anchor"></a>Including annotations</h2>
<p>Annotations are stored in <code>clinical.data</code> slot of MAF.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="../reference/getClinicalData.html">getClinicalData</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">laml</span>)
<span class="co">#&gt;      Tumor_Sample_Barcode FAB_classification days_to_last_followup</span>
<span class="co">#&gt;   1:         TCGA-AB-2802                 M4                   365</span>
<span class="co">#&gt;   2:         TCGA-AB-2803                 M3                   792</span>
<span class="co">#&gt;   3:         TCGA-AB-2804                 M3                  2557</span>
<span class="co">#&gt;   4:         TCGA-AB-2805                 M0                   577</span>
<span class="co">#&gt;   5:         TCGA-AB-2806                 M1                   945</span>
<span class="co">#&gt;  ---                                                              </span>
<span class="co">#&gt; 189:         TCGA-AB-3007                 M3                  1581</span>
<span class="co">#&gt; 190:         TCGA-AB-3008                 M1                   822</span>
<span class="co">#&gt; 191:         TCGA-AB-3009                 M4                   577</span>
<span class="co">#&gt; 192:         TCGA-AB-3011                 M1                  1885</span>
<span class="co">#&gt; 193:         TCGA-AB-3012                 M3                  1887</span>
<span class="co">#&gt;      Overall_Survival_Status</span>
<span class="co">#&gt;   1:                       1</span>
<span class="co">#&gt;   2:                       1</span>
<span class="co">#&gt;   3:                       0</span>
<span class="co">#&gt;   4:                       1</span>
<span class="co">#&gt;   5:                       1</span>
<span class="co">#&gt;  ---                        </span>
<span class="co">#&gt; 189:                       0</span>
<span class="co">#&gt; 190:                       1</span>
<span class="co">#&gt; 191:                       1</span>
<span class="co">#&gt; 192:                       0</span>
<span class="co">#&gt; 193:                       0</span></pre></body></html></div>
<p>Include <code>FAB_classification</code> from clinical data as one of the sample annotations.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="../reference/oncoplot.html">oncoplot</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">genes</span> <span class="kw">=</span> <span class="no">aml_genes</span>, <span class="kw">clinicalFeatures</span> <span class="kw">=</span> <span class="st">'FAB_classification'</span>)</pre></body></html></div>
<p><img src="oncoplots_files/figure-html/unnamed-chunk-10-1.png" width="960" style="display: block; margin: auto auto auto 0;"></p>
<p>More than one annotations can be included by passing them to the argument <code>clinicalFeatures</code>. Above plot can be further enhanced by sorting according to annotations. Custom colors can be specified as a list of named vectors for each levels.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co">#Color coding for FAB classification</span>
<span class="no">fabcolors</span> <span class="kw">=</span> <span class="kw pkg">RColorBrewer</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span>(<span class="kw">n</span> <span class="kw">=</span> <span class="fl">8</span>,<span class="kw">name</span> <span class="kw">=</span> <span class="st">'Spectral'</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">fabcolors</span>) <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"M0"</span>, <span class="st">"M1"</span>, <span class="st">"M2"</span>, <span class="st">"M3"</span>, <span class="st">"M4"</span>, <span class="st">"M5"</span>, <span class="st">"M6"</span>, <span class="st">"M7"</span>)
<span class="no">fabcolors</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">FAB_classification</span> <span class="kw">=</span> <span class="no">fabcolors</span>)

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="no">fabcolors</span>)
<span class="co">#&gt; $FAB_classification</span>
<span class="co">#&gt;        M0        M1        M2        M3        M4        M5        M6        M7 </span>
<span class="co">#&gt; "#D53E4F" "#F46D43" "#FDAE61" "#FEE08B" "#E6F598" "#ABDDA4" "#66C2A5" "#3288BD"</span>

<span class="fu"><a href="../reference/oncoplot.html">oncoplot</a></span>(
  <span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">genes</span> <span class="kw">=</span> <span class="no">aml_genes</span>,
  <span class="kw">clinicalFeatures</span> <span class="kw">=</span> <span class="st">'FAB_classification'</span>,
  <span class="kw">sortByAnnotation</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
  <span class="kw">annotationColor</span> <span class="kw">=</span> <span class="no">fabcolors</span>
)</pre></body></html></div>
<p><img src="oncoplots_files/figure-html/unnamed-chunk-11-1.png" width="960" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="highlighting-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#highlighting-samples" class="anchor"></a>Highlighting samples</h2>
<p>If you prefer to highlight mutations by a specific attribute, you can use <code>additionalFeature</code> argument.</p>
<p>Example: Highlight all mutations where alt allele is C.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="../reference/oncoplot.html">oncoplot</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">genes</span> <span class="kw">=</span> <span class="no">aml_genes</span>,
         <span class="kw">additionalFeature</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Tumor_Seq_Allele2"</span>, <span class="st">"C"</span>))</pre></body></html></div>
<p><img src="oncoplots_files/figure-html/unnamed-chunk-12-1.png" width="960" style="display: block; margin: auto auto auto 0;"></p>
<p>Note that first argument (Tumor_Seq_Allele2) must a be column in MAF file, and second argument (C) is a value in that column. If you want to know what columns are present in the MAF file, use <code>getFields</code>.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="fu"><a href="../reference/getFields.html">getFields</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">laml</span>)
<span class="co">#&gt;  [1] "Hugo_Symbol"            "Entrez_Gene_Id"         "Center"                </span>
<span class="co">#&gt;  [4] "NCBI_Build"             "Chromosome"             "Start_Position"        </span>
<span class="co">#&gt;  [7] "End_Position"           "Strand"                 "Variant_Classification"</span>
<span class="co">#&gt; [10] "Variant_Type"           "Reference_Allele"       "Tumor_Seq_Allele1"     </span>
<span class="co">#&gt; [13] "Tumor_Seq_Allele2"      "Tumor_Sample_Barcode"   "Protein_Change"        </span>
<span class="co">#&gt; [16] "i_TumorVAF_WU"          "i_transcript_name"</span></pre></body></html></div>
</div>
<div id="group-by-pathways" class="section level2">
<h2 class="hasAnchor">
<a href="#group-by-pathways" class="anchor"></a>Group by Pathways</h2>
<p>Genes can be auto grouped based on their Biological processess by setting <code>pathways = 'auto'</code> or by providing custom pathway list in the form of a two column tsv file or a data.frame containing gene names and their corresponding pathway.</p>
<div id="auto" class="section level3">
<h3 class="hasAnchor">
<a href="#auto" class="anchor"></a>Auto</h3>
<p>setting <code>pathways = 'auto'</code> draws top 3 most affected pathways</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="../reference/oncoplot.html">oncoplot</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">pathways</span> <span class="kw">=</span> <span class="st">"auto"</span>, <span class="kw">gene_mar</span> <span class="kw">=</span> <span class="fl">8</span>, <span class="kw">fontSize</span> <span class="kw">=</span> <span class="fl">0.6</span>)
<span class="co">#&gt; Drawing upto top 3 mutated pathways</span>
<span class="co">#&gt;                                Pathway  N n_affected_genes fraction_affected</span>
<span class="co">#&gt;  1:                      RTK_signaling 16                6        0.37500000</span>
<span class="co">#&gt;  2:          Epigenetics_DNA_modifiers  1                1        1.00000000</span>
<span class="co">#&gt;  3:                         Metabolism  2                2        1.00000000</span>
<span class="co">#&gt;  4:               Transcription_factor 39               11        0.28205128</span>
<span class="co">#&gt;  5:                   Genome_integrity 14                5        0.35714286</span>
<span class="co">#&gt;  6:                     MAPK_signaling  9                3        0.33333333</span>
<span class="co">#&gt;  7:                    Other_signaling 28                5        0.17857143</span>
<span class="co">#&gt;  8:                           Splicing  6                3        0.50000000</span>
<span class="co">#&gt;  9:               Histone_modification  3                2        0.66666667</span>
<span class="co">#&gt; 10:                              Other 22                3        0.13636364</span>
<span class="co">#&gt; 11: Protein_homeostasis/ubiquitination 15               NA                NA</span>
<span class="co">#&gt; 12:                   Immune_signaling 10                3        0.30000000</span>
<span class="co">#&gt; 13:                    Chromatin_other 14                7        0.50000000</span>
<span class="co">#&gt; 14:                    NOTCH_signaling  1                1        1.00000000</span>
<span class="co">#&gt; 15:                     PI3K_signaling  9                1        0.11111111</span>
<span class="co">#&gt; 16:                      RNA_abundance 15                1        0.06666667</span>
<span class="co">#&gt; 17:          Chromatin_SWI/SNF_complex  8               NA                NA</span>
<span class="co">#&gt; 18:            Wnt/B-catenin_signaling  8               NA                NA</span>
<span class="co">#&gt;     Mutated_samples Fraction_mutated_samples</span>
<span class="co">#&gt;  1:              63              0.326424870</span>
<span class="co">#&gt;  2:              48              0.248704663</span>
<span class="co">#&gt;  3:              37              0.191709845</span>
<span class="co">#&gt;  4:              35              0.181347150</span>
<span class="co">#&gt;  5:              28              0.145077720</span>
<span class="co">#&gt;  6:              25              0.129533679</span>
<span class="co">#&gt;  7:              13              0.067357513</span>
<span class="co">#&gt;  8:              11              0.056994819</span>
<span class="co">#&gt;  9:               4              0.020725389</span>
<span class="co">#&gt; 10:               4              0.020725389</span>
<span class="co">#&gt; 11:               4              0.020725389</span>
<span class="co">#&gt; 12:               3              0.015544041</span>
<span class="co">#&gt; 13:               2              0.010362694</span>
<span class="co">#&gt; 14:               1              0.005181347</span>
<span class="co">#&gt; 15:               1              0.005181347</span>
<span class="co">#&gt; 16:               1              0.005181347</span>
<span class="co">#&gt; 17:               0              0.000000000</span>
<span class="co">#&gt; 18:               0              0.000000000</span></pre></body></html></div>
<p><img src="oncoplots_files/figure-html/unnamed-chunk-14-1.png" width="960" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="custom-pathways" class="section level3">
<h3 class="hasAnchor">
<a href="#custom-pathways" class="anchor"></a>Custom pathways</h3>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">pathways</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(
  <span class="kw">Genes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(
    <span class="st">"TP53"</span>,
    <span class="st">"WT1"</span>,
    <span class="st">"PHF6"</span>,
    <span class="st">"DNMT3A"</span>,
    <span class="st">"DNMT3B"</span>,
    <span class="st">"TET1"</span>,
    <span class="st">"TET2"</span>,
    <span class="st">"IDH1"</span>,
    <span class="st">"IDH2"</span>,
    <span class="st">"FLT3"</span>,
    <span class="st">"KIT"</span>,
    <span class="st">"KRAS"</span>,
    <span class="st">"NRAS"</span>,
    <span class="st">"RUNX1"</span>,
    <span class="st">"CEBPA"</span>,
    <span class="st">"ASXL1"</span>,
    <span class="st">"EZH2"</span>,
    <span class="st">"KDM6A"</span>
  ),
  <span class="kw">Pathway</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(
    <span class="st">"TSG"</span>, <span class="st">"DNAm"</span>, <span class="st">"Signalling"</span>, <span class="st">"TFs"</span>, <span class="st">"ChromMod"</span>
  ), <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">3</span>)),
  <span class="kw">stringsAsFactors</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">pathways</span>)
<span class="co">#&gt;    Genes Pathway</span>
<span class="co">#&gt; 1   TP53     TSG</span>
<span class="co">#&gt; 2    WT1     TSG</span>
<span class="co">#&gt; 3   PHF6     TSG</span>
<span class="co">#&gt; 4 DNMT3A    DNAm</span>
<span class="co">#&gt; 5 DNMT3B    DNAm</span>
<span class="co">#&gt; 6   TET1    DNAm</span>

<span class="fu"><a href="../reference/oncoplot.html">oncoplot</a></span>(<span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml</span>, <span class="kw">pathways</span> <span class="kw">=</span> <span class="no">pathways</span>, <span class="kw">gene_mar</span> <span class="kw">=</span> <span class="fl">8</span>, <span class="kw">fontSize</span> <span class="kw">=</span> <span class="fl">0.6</span>)</pre></body></html></div>
<p><img src="oncoplots_files/figure-html/unnamed-chunk-15-1.png" width="960" style="display: block; margin: auto auto auto 0;"></p>
</div>
</div>
<div id="combining-everything" class="section level2">
<h2 class="hasAnchor">
<a href="#combining-everything" class="anchor"></a>Combining everything</h2>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu"><a href="../reference/oncoplot.html">oncoplot</a></span>(
  <span class="kw">maf</span> <span class="kw">=</span> <span class="no">laml.plus.gistic</span>,
  <span class="kw">draw_titv</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
  <span class="kw">pathways</span> <span class="kw">=</span> <span class="no">pathways</span>,
  <span class="kw">clinicalFeatures</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'FAB_classification'</span>, <span class="st">'Overall_Survival_Status'</span>),
  <span class="kw">sortByAnnotation</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
  <span class="kw">additionalFeature</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Tumor_Seq_Allele2"</span>, <span class="st">"C"</span>),
  <span class="kw">leftBarData</span> <span class="kw">=</span> <span class="no">aml_genes_vaf</span>,
  <span class="kw">leftBarLims</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">100</span>),
  <span class="kw">rightBarData</span> <span class="kw">=</span> <span class="no">laml.mutsig</span>[,<span class="fu">.</span>(<span class="no">gene</span>, <span class="no">q</span>)],
)</pre></body></html></div>
<p><img src="oncoplots_files/figure-html/unnamed-chunk-16-1.png" width="960" style="display: block; margin: auto auto auto 0;"></p>
</div>
<div id="sessioninfo" class="section level2">
<h2 class="hasAnchor">
<a href="#sessioninfo" class="anchor"></a>SessionInfo</h2>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 4.0.2 (2020-06-22)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.6</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] maftools_2.4.10</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Rcpp_1.0.5         rstudioapi_0.11    knitr_1.29         magrittr_1.5      </span>
<span class="co">#&gt;  [5] splines_4.0.2      MASS_7.3-51.6      lattice_0.20-41    R6_2.4.1          </span>
<span class="co">#&gt;  [9] rlang_0.4.7        stringr_1.4.0      tools_4.0.2        grid_4.0.2        </span>
<span class="co">#&gt; [13] data.table_1.13.0  xfun_0.16          R.oo_1.23.0        htmltools_0.5.0   </span>
<span class="co">#&gt; [17] yaml_2.2.1         survival_3.2-3     assertthat_0.2.1   digest_0.6.25     </span>
<span class="co">#&gt; [21] rprojroot_1.3-2    pkgdown_1.5.1      crayon_1.3.4       Matrix_1.2-18     </span>
<span class="co">#&gt; [25] RColorBrewer_1.1-2 R.utils_2.9.2      fs_1.5.0           memoise_1.1.0     </span>
<span class="co">#&gt; [29] evaluate_0.14      rmarkdown_2.3      stringi_1.4.6      compiler_4.0.2    </span>
<span class="co">#&gt; [33] R.methodsS3_1.8.0  desc_1.2.0         backports_1.1.8</span></pre></body></html></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Anand Mayakonda.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
