<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ maftools</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maftools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.4.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/maftools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/oncoplots.html">Customizing oncoplots</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/PoisonAlien/maftools/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data-input-classes-and-operation" class="hasAnchor"><a href="#section-data-input-classes-and-operation" class="anchor"></a>Data input, classes and operation</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="MAF-class.html">MAF-class</a></code> </p>
        </td>
        <td><p>Class MAF</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="GISTIC-class.html">GISTIC-class</a></code> </p>
        </td>
        <td><p>Class GISTIC</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read.maf.html">read.maf()</a></code> </p>
        </td>
        <td><p>Read MAF files.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="readGistic.html">readGistic()</a></code> </p>
        </td>
        <td><p>Read and summarize gistic output.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annovarToMaf.html">annovarToMaf()</a></code> </p>
        </td>
        <td><p>Converts annovar annotations into MAF.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="icgcSimpleMutationToMAF.html">icgcSimpleMutationToMAF()</a></code> </p>
        </td>
        <td><p>Converts ICGC Simple Somatic Mutation format file to MAF</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subsetMaf.html">subsetMaf()</a></code> </p>
        </td>
        <td><p>Subset MAF objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterMaf.html">filterMaf()</a></code> </p>
        </td>
        <td><p>Filter MAF objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="merge_mafs.html">merge_mafs()</a></code> </p>
        </td>
        <td><p>Merge multiple mafs into single MAF</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="setMaf.html">setdiffMAF()</a></code> <code><a href="setMaf.html">intersectMAF()</a></code> </p>
        </td>
        <td><p>Set Operations for MAF objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write.mafSummary.html">write.mafSummary()</a></code> </p>
        </td>
        <td><p>Writes maf summaries to output tab-delimited text files.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write.GisticSummary.html">write.GisticSummary()</a></code> </p>
        </td>
        <td><p>Writes GISTIC summaries to output tab-delimited text files.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-analyze-and-visualize-maf-object" class="hasAnchor"><a href="#section-analyze-and-visualize-maf-object" class="anchor"></a>Analyze and visualize MAF object</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="plotmafSummary.html">plotmafSummary()</a></code> </p>
        </td>
        <td><p>Plots maf summary.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotVaf.html">plotVaf()</a></code> </p>
        </td>
        <td><p>Plots vaf distribution of genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotTiTv.html">plotTiTv()</a></code> </p>
        </td>
        <td><p>Plot Transition and Trasnversion ratios.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rainfallPlot.html">rainfallPlot()</a></code> </p>
        </td>
        <td><p>Rainfall plot to display hyper mutated genomic regions.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lollipopPlot.html">lollipopPlot()</a></code> </p>
        </td>
        <td><p>Draws lollipop plot of amino acid changes on to Protein structure.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="oncoplot.html">oncoplot()</a></code> </p>
        </td>
        <td><p>draw an oncoplot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="oncostrip.html">oncostrip()</a></code> </p>
        </td>
        <td><p>draw an oncostrip similar to cBioportal oncoprinter output.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="oncodrive.html">oncodrive()</a></code> </p>
        </td>
        <td><p>Detect cancer driver genes based on positional clustering of variants.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotOncodrive.html">plotOncodrive()</a></code> </p>
        </td>
        <td><p>Plots results from <code>oncodrive</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pfamDomains.html">pfamDomains()</a></code> </p>
        </td>
        <td><p>pfam domain annotation and summarization.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="somaticInteractions.html">somaticInteractions()</a></code> </p>
        </td>
        <td><p>Exact tests to detect mutually exclusive, co-occuring and altered genesets.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="OncogenicPathways.html">OncogenicPathways()</a></code> </p>
        </td>
        <td><p>Enrichment of known oncogenic pathways</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PlotOncogenicPathways.html">PlotOncogenicPathways()</a></code> </p>
        </td>
        <td><p>Plot oncogenic pathways</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="inferHeterogeneity.html">inferHeterogeneity()</a></code> </p>
        </td>
        <td><p>Clusters variants based on Variant Allele Frequencies (VAF).</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotClusters.html">plotClusters()</a></code> </p>
        </td>
        <td><p>Plot density plots from clutering results.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="math.score.html">math.score()</a></code> </p>
        </td>
        <td><p>calculates MATH (Mutant-Allele Tumor Heterogeneity) score.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="forestPlot.html">forestPlot()</a></code> </p>
        </td>
        <td><p>Draw forest plot for differences betweeen cohorts.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="drugInteractions.html">drugInteractions()</a></code> </p>
        </td>
        <td><p>Drug-Gene Interactions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mafSurvival.html">mafSurvival()</a></code> </p>
        </td>
        <td><p>Performs survival analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="survGroup.html">survGroup()</a></code> </p>
        </td>
        <td><p>Predict genesets associated with survival</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mafSurvGroup.html">mafSurvGroup()</a></code> </p>
        </td>
        <td><p>Performs survival analysis for a geneset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clinicalEnrichment.html">clinicalEnrichment()</a></code> </p>
        </td>
        <td><p>Performs mutational enrichment analysis for a given clinical feature.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotEnrichmentResults.html">plotEnrichmentResults()</a></code> </p>
        </td>
        <td><p>Plots results from clinicalEnrichment analysis</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-analyze-and-visualize-gistic-object" class="hasAnchor"><a href="#section-analyze-and-visualize-gistic-object" class="anchor"></a>Analyze and visualize GISTIC object</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="plotCBSsegments.html">plotCBSsegments()</a></code> </p>
        </td>
        <td><p>Plots segmented copy number data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gisticOncoPlot.html">gisticOncoPlot()</a></code> </p>
        </td>
        <td><p>Plot gistic results.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gisticBubblePlot.html">gisticBubblePlot()</a></code> </p>
        </td>
        <td><p>Plot gistic results as a bubble plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gisticChromPlot.html">gisticChromPlot()</a></code> </p>
        </td>
        <td><p>Plot gistic results along linearized chromosome</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-analyze-and-visualize-multiple-cohorts" class="hasAnchor"><a href="#section-analyze-and-visualize-multiple-cohorts" class="anchor"></a>Analyze and visualize multiple cohorts</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="mafCompare.html">mafCompare()</a></code> </p>
        </td>
        <td><p>compare two cohorts (MAF).</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vafCompare.html">vafCompare()</a></code> </p>
        </td>
        <td><p>compare VAF across two cohorts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="coBarplot.html">coBarplot()</a></code> </p>
        </td>
        <td><p>Draw two barplots side by side for cohort comparision.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="coOncoplot.html">coOncoplot()</a></code> </p>
        </td>
        <td><p>Draw two oncoplots side by side for cohort comparision.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lollipopPlot2.html">lollipopPlot2()</a></code> </p>
        </td>
        <td><p>Compare two lollipop plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tcgaCompare.html">tcgaCompare()</a></code> </p>
        </td>
        <td><p>Compare mutation load against TCGA cohorts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tcgaDriverBP.html">tcgaDriverBP()</a></code> </p>
        </td>
        <td><p>Compare genes to known TCGA drivers and their biological pathways</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-apobec-and-mutational-signature-analysis" class="hasAnchor"><a href="#section-apobec-and-mutational-signature-analysis" class="anchor"></a>APOBEC and mutational signature analysis</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="trinucleotideMatrix.html">trinucleotideMatrix()</a></code> </p>
        </td>
        <td><p>Extract single 5' and 3' bases flanking the mutated site for de-novo signature analysis. Also estimates APOBEC enrichment scores.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotApobecDiff.html">plotApobecDiff()</a></code> </p>
        </td>
        <td><p>Plot differences between APOBEC enriched and non-APOBEC enriched samples.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepareMutSig.html">prepareMutSig()</a></code> </p>
        </td>
        <td><p>Prepares MAF file for MutSig analysis.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="estimateSignatures.html">estimateSignatures()</a></code> </p>
        </td>
        <td><p>Estimate number of signatures based on cophenetic correlation metric</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCophenetic.html">plotCophenetic()</a></code> </p>
        </td>
        <td><p>Draw an elbow plot of cophenetic correlation metric.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extractSignatures.html">extractSignatures()</a></code> </p>
        </td>
        <td><p>Extract mutational signatures from trinucleotide context.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compareSignatures.html">compareSignatures()</a></code> </p>
        </td>
        <td><p>Compares identified denovo mutational signatures to known COSMIC signatures</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSignatures.html">plotSignatures()</a></code> </p>
        </td>
        <td><p>Plots decomposed mutational signatures</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="signatureEnrichment.html">signatureEnrichment()</a></code> </p>
        </td>
        <td><p>Performs sample stratification based on signature contribution and enrichment analysis.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-obtain-information-from-maf-or-gistic-object" class="hasAnchor"><a href="#section-obtain-information-from-maf-or-gistic-object" class="anchor"></a>Obtain information from MAF or GISTIC object</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="mafSummary.html">mafSummary()</a></code> </p>
        </td>
        <td><p>Summary statistics of MAF</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getClinicalData.html">getClinicalData()</a></code> </p>
        </td>
        <td><p>extract annotations from MAF object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGeneSummary.html">getGeneSummary()</a></code> </p>
        </td>
        <td><p>extract gene summary from MAF or GISTIC object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSampleSummary.html">getSampleSummary()</a></code> </p>
        </td>
        <td><p>extract sample summary from MAF or GISTIC object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getFields.html">getFields()</a></code> </p>
        </td>
        <td><p>extract available fields from MAF object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getCytobandSummary.html">getCytobandSummary()</a></code> </p>
        </td>
        <td><p>extract cytoband summary from GISTIC object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genesToBarcodes.html">genesToBarcodes()</a></code> </p>
        </td>
        <td><p>Extracts Tumor Sample Barcodes where the given genes are mutated.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genotypeMatrix.html">genotypeMatrix()</a></code> </p>
        </td>
        <td><p>Creates a Genotype Matrix for every variant</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tmb.html">tmb()</a></code> </p>
        </td>
        <td><p>Estimate Tumor Mutation Burden</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="titv.html">titv()</a></code> </p>
        </td>
        <td><p>Classifies SNPs into transitions and transversions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mutCountMatrix.html">mutCountMatrix()</a></code> </p>
        </td>
        <td><p>Generates count matrix of mutations.</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Anand Mayakonda.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


