<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Clusters variants based on Variant Allele Frequencies (VAF). — inferHeterogeneity • maftools</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Clusters variants based on Variant Allele Frequencies (VAF). — inferHeterogeneity" />
<meta property="og:description" content="takes output generated by read.maf and clusters variants to infer tumor heterogeneity. This function requires VAF for clustering and density estimation.
VAF can be on the scale 0-1 or 0-100. Optionally if copy number information is available, it can be provided as a segmented file (e.g, from Circular Binary Segmentation). Those variants in
copy number altered regions will be ignored." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maftools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.4.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/maftools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/oncoplots.html">Customizing oncoplots</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/PoisonAlien/maftools/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Clusters variants based on Variant Allele Frequencies (VAF).</h1>
    <small class="dont-index">Source: <a href='https://github.com/PoisonAlien/maftools/blob/master/R/inferTumHetero.R'><code>R/inferTumHetero.R</code></a></small>
    <div class="hidden name"><code>inferHeterogeneity.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>takes output generated by read.maf and clusters variants to infer tumor heterogeneity. This function requires VAF for clustering and density estimation.
VAF can be on the scale 0-1 or 0-100. Optionally if copy number information is available, it can be provided as a segmented file (e.g, from Circular Binary Segmentation). Those variants in
copy number altered regions will be ignored.</p>
    </div>

    <pre class="usage"><span class='fu'>inferHeterogeneity</span>(
  <span class='no'>maf</span>,
  <span class='kw'>tsb</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>top</span> <span class='kw'>=</span> <span class='fl'>5</span>,
  <span class='kw'>vafCol</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>segFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>ignChr</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>minVaf</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>maxVaf</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>useSyn</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>dirichlet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>maf</th>
      <td><p>an <code><a href='MAF-class.html'>MAF</a></code> object generated by <code><a href='read.maf.html'>read.maf</a></code></p></td>
    </tr>
    <tr>
      <th>tsb</th>
      <td><p>specify sample names (Tumor_Sample_Barcodes) for which clustering has to be done.</p></td>
    </tr>
    <tr>
      <th>top</th>
      <td><p>if <code>tsb</code> is NULL, uses top n number of most mutated samples. Defaults to 5.</p></td>
    </tr>
    <tr>
      <th>vafCol</th>
      <td><p>manually specify column name for vafs. Default looks for column 't_vaf'</p></td>
    </tr>
    <tr>
      <th>segFile</th>
      <td><p>path to CBS segmented copy number file. Column names should be Sample, Chromosome, Start, End, Num_Probes and Segment_Mean (log2 scale).</p></td>
    </tr>
    <tr>
      <th>ignChr</th>
      <td><p>ignore these chromosomes from analysis. e.g, sex chromsomes chrX, chrY. Default NULL.</p></td>
    </tr>
    <tr>
      <th>minVaf</th>
      <td><p>filter low frequency variants. Low vaf variants maybe due to sequencing error. Default 0. (on the scale of 0 to 1)</p></td>
    </tr>
    <tr>
      <th>maxVaf</th>
      <td><p>filter high frequency variants. High vaf variants maybe due to copy number alterations or impure tumor. Default 1. (on the scale of 0 to 1)</p></td>
    </tr>
    <tr>
      <th>useSyn</th>
      <td><p>Use synonymous variants. Default FALSE.</p></td>
    </tr>
    <tr>
      <th>dirichlet</th>
      <td><p>Deprecated! No longer supported. uses nonparametric dirichlet process for clustering. Default FALSE - uses finite mixture models.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>list of clustering tables.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function clusters variants based on VAF to estimate univariate density and cluster classification. There are two methods available
for clustering. Default using parametric finite mixture models and another method using nonparametric inifinite mixture models (Dirichlet process).</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Chris Fraley and Adrian E. Raftery (2002) Model-based Clustering, Discriminant Analysis and Density Estimation Journal of the American
Statistical Association 97:611-631</p>
<p>Jara A, Hanson TE, Quintana FA, Muller P, Rosner GL. DPpackage: Bayesian Semi- and Nonparametric Modeling in R. Journal of statistical software. 2011;40(5):1-30.</p>
<p>Olshen AB, Venkatraman ES, Lucito R, Wigler M. Circular binary segmentation for the analysis of array-based DNA copy number data. Biostatistics. 2004;5(4):557-72.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='plotClusters.html'>plotClusters</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='no'>laml.maf</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"tcga_laml.maf.gz"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"maftools"</span>)
<span class='no'>laml</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='read.maf.html'>read.maf</a></span>(<span class='kw'>maf</span> <span class='kw'>=</span> <span class='no'>laml.maf</span>)
<span class='no'>TCGA.AB.2972.clust</span> <span class='kw'>&lt;-</span> <span class='fu'>inferHeterogeneity</span>(<span class='kw'>maf</span> <span class='kw'>=</span> <span class='no'>laml</span>, <span class='kw'>tsb</span> <span class='kw'>=</span> <span class='st'>'TCGA-AB-2972'</span>, <span class='kw'>vafCol</span> <span class='kw'>=</span> <span class='st'>'i_TumorVAF_WU'</span>)
}</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Anand Mayakonda.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


