<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Rainfall plot to display hyper mutated genomic regions. — rainfallPlot • maftools</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Rainfall plot to display hyper mutated genomic regions. — rainfallPlot" />
<meta property="og:description" content="Plots inter variant distance as a function of genomic locus." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maftools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.4.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/maftools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/oncoplots.html">Customizing oncoplots</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/PoisonAlien/maftools/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Rainfall plot to display hyper mutated genomic regions.</h1>
    <small class="dont-index">Source: <a href='https://github.com/PoisonAlien/maftools/blob/master/R/rainfallPlot.R'><code>R/rainfallPlot.R</code></a></small>
    <div class="hidden name"><code>rainfallPlot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plots inter variant distance as a function of genomic locus.</p>
    </div>

    <pre class="usage"><span class='fu'>rainfallPlot</span>(
  <span class='no'>maf</span>,
  <span class='kw'>tsb</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>detectChangePoints</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>ref.build</span> <span class='kw'>=</span> <span class='st'>"hg19"</span>,
  <span class='kw'>color</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>savePlot</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>width</span> <span class='kw'>=</span> <span class='fl'>6</span>,
  <span class='kw'>height</span> <span class='kw'>=</span> <span class='fl'>3</span>,
  <span class='kw'>fontSize</span> <span class='kw'>=</span> <span class='fl'>1.2</span>,
  <span class='kw'>pointSize</span> <span class='kw'>=</span> <span class='fl'>0.4</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>maf</th>
      <td><p>an <code><a href='MAF-class.html'>MAF</a></code> object generated by <code><a href='read.maf.html'>read.maf</a></code>. Required.</p></td>
    </tr>
    <tr>
      <th>tsb</th>
      <td><p>specify sample names (Tumor_Sample_Barcodes) for which plotting has to be done. If NULL, draws plot for most mutated sample.</p></td>
    </tr>
    <tr>
      <th>detectChangePoints</th>
      <td><p>If TRUE, detectes genomic change points where potential kataegis are formed. Results are written to an output tab delimted file.</p></td>
    </tr>
    <tr>
      <th>ref.build</th>
      <td><p>Reference build for chromosome sizes. Can be hg18, hg19 or hg38. Default hg19.</p></td>
    </tr>
    <tr>
      <th>color</th>
      <td><p>named vector of colors for each coversion class.</p></td>
    </tr>
    <tr>
      <th>savePlot</th>
      <td><p>If TRUE plot is saved to output pdf. Default FALSE.</p></td>
    </tr>
    <tr>
      <th>width</th>
      <td><p>width of plot to be saved.</p></td>
    </tr>
    <tr>
      <th>height</th>
      <td><p>height of plot to be saved.</p></td>
    </tr>
    <tr>
      <th>fontSize</th>
      <td><p>Default 12.</p></td>
    </tr>
    <tr>
      <th>pointSize</th>
      <td><p>Default 0.8.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Results are written to an output file with suffix changePoints.tsv</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>If `detectChangePoints`` is set to TRUE, this function will identify Kataegis loci.
Kategis detection algorithm by Moritz Goretzky at WWU Munster, which exploits the definition of Kategis (six consecutive mutations with an avg. distance of 1000bp ) to idetify hyper mutated genomic loci.
Algorithm starts with a double-ended queue to which six consecutive mutations are added and their average intermutation distance is calculated.
If the average intermutation distance is larger than 1000, one element is added at the back of the queue and one is removed from the front.
If the average intermutation distance is less or equal to 1000, further mutations are added until the average intermutation distance is larger than 1000.
After that all mutations in the double-ended queue are written into output as one kataegis and the double-ended queue is reinitialized with six mutations.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Anand Mayakonda.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


