<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Extract single 5' and 3' bases flanking the mutated site for de-novo signature analysis. Also estimates APOBEC enrichment scores. — trinucleotideMatrix • maftools</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Extract single 5' and 3' bases flanking the mutated site for de-novo signature analysis. Also estimates APOBEC enrichment scores. — trinucleotideMatrix" />
<meta property="og:description" content="Extract single 5' and 3' bases flanking the mutated site for de-novo signature analysis. Also estimates APOBEC enrichment scores." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">maftools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.4.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/maftools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/oncoplots.html">Customizing oncoplots</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/PoisonAlien/maftools/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Extract single 5' and 3' bases flanking the mutated site for de-novo signature analysis. Also estimates APOBEC enrichment scores.</h1>
    <small class="dont-index">Source: <a href='https://github.com/PoisonAlien/maftools/blob/master/R/TrinucleotideMatrix.R'><code>R/TrinucleotideMatrix.R</code></a></small>
    <div class="hidden name"><code>trinucleotideMatrix.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Extract single 5' and 3' bases flanking the mutated site for de-novo signature analysis. Also estimates APOBEC enrichment scores.</p>
    </div>

    <pre class="usage"><span class='fu'>trinucleotideMatrix</span>(
  <span class='no'>maf</span>,
  <span class='kw'>ref_genome</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>prefix</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>add</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>ignoreChr</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>useSyn</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>fn</span> <span class='kw'>=</span> <span class='kw'>NULL</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>maf</th>
      <td><p>an <code><a href='MAF-class.html'>MAF</a></code> object generated by <code><a href='read.maf.html'>read.maf</a></code></p></td>
    </tr>
    <tr>
      <th>ref_genome</th>
      <td><p>BSgenome object or name of the installed BSgenome package. Example: BSgenome.Hsapiens.UCSC.hg19
Default NULL, tries to auto-detect from installed genomes.</p></td>
    </tr>
    <tr>
      <th>prefix</th>
      <td><p>Prefix to add or remove from contig names in MAF file.</p></td>
    </tr>
    <tr>
      <th>add</th>
      <td><p>If prefix is used, default is to add prefix to contig names in MAF file. If false prefix will be removed from contig names.</p></td>
    </tr>
    <tr>
      <th>ignoreChr</th>
      <td><p>Chromsomes to ignore from analysis. e.g. chrM</p></td>
    </tr>
    <tr>
      <th>useSyn</th>
      <td><p>Logical. Whether to include synonymous variants in analysis. Defaults to TRUE</p></td>
    </tr>
    <tr>
      <th>fn</th>
      <td><p>If given writes APOBEC results to an output file with basename fn. Default NULL.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>list of 2. A matrix of dimension nx96, where n is the number of samples in the MAF and a table describing APOBEC enrichment per sample.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Extracts immediate 5' and 3' bases flanking the mutated site and classifies them into 96 substitution classes.
Requires BSgenome data packages for sequence extraction.</p>
<p>APOBEC Enrichment: Enrichment score is calculated using the same method described by Roberts et al.</p>
<p>E = (n_tcw * background_c) / (n_C * background_tcw)</p>
<p>where, n_tcw = number of mutations within T[C&gt;T]W and T[C&gt;G]W context. (W -&gt; A or T)</p>
<p>n_C   = number of mutated C and G</p>
<p>background_C and background_tcw motifs are number of C and TCW motifs occuring around +/- 20bp of each mutation.</p>
<p>One-sided Fisher's Exact test is performed to determine the enrichment of APOBEC tcw mutations over background.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Roberts SA, Lawrence MS, Klimczak LJ, et al. An APOBEC Cytidine Deaminase Mutagenesis Pattern is Widespread in Human Cancers. Nature genetics. 2013;45(9):970-976. doi:10.1038/ng.2702.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='extractSignatures.html'>extractSignatures</a></code> <code><a href='plotApobecDiff.html'>plotApobecDiff</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='no'>laml.maf</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"tcga_laml.maf.gz"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"maftools"</span>)
<span class='no'>laml</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='read.maf.html'>read.maf</a></span>(<span class='kw'>maf</span> <span class='kw'>=</span> <span class='no'>laml.maf</span>)
<span class='no'>laml.tnm</span> <span class='kw'>&lt;-</span> <span class='fu'>trinucleotideMatrix</span>(<span class='kw'>maf</span> <span class='kw'>=</span> <span class='no'>laml</span>, <span class='kw'>ref_genome</span> <span class='kw'>=</span> <span class='st'>'BSgenome.Hsapiens.UCSC.hg19'</span>,
<span class='kw'>prefix</span> <span class='kw'>=</span> <span class='st'>'chr'</span>, <span class='kw'>add</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>useSyn</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
}</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Anand Mayakonda.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


