#' Draws lollipop plot of amino acid changes on to Protein structure.
#'
#' @description Draws lollipop plot of amino acid changes.
#' @details This function by default looks for field "AAChange" in maf file. One can also manually specify field name containing amino acid changes.
#' @param maf an \code{\link{MAF}} object generated by \code{\link{read.maf}}
#' @param gene HGNC symbol for which protein structure to be drawn.
#' @param refSeqID RefSeq transcript identifier for \code{gene} if known.
#' @param proteinID RefSeq protein identifier for \code{gene} if known.
#' @param labelPos Amino acid positions to label. If 'all', labels all variants.
#' @param repel If points are too close to each other, use this option to repel them. Default FALSE. Warning: naive method, might make plot ugly in case of too many variants!
#' @param AACol manually specify column name for amino acid changes. Default looks for field 'AAChange'. Changes can be of any format i.e, can be a numeric value or HGVSp annotations (e.g; p.P459L, p.L2195Pfs*30 or p.Leu2195ProfsTer30)
#' @param colors named vector of colors for each Variant_Classification. Default NULL.
#' @param legendTxtSize Text size for legend. Default 10
#' @return ggplot object of the plot, which can be futher modified.
#' @import ggrepel
#' @examples
#' laml.maf <- system.file("extdata", "tcga_laml.maf.gz", package = "maftools")
#' laml <- read.maf(maf = laml.maf, removeSilent = TRUE, useAll = FALSE)
#' lollipopPlot(maf = laml, gene = 'KIT', AACol = 'Protein_Change')
#'
#' @export


lollipopPlot = function(maf, gene = NULL, refSeqID = NULL, proteinID = NULL, labelPos = NULL, AACol = NULL, repel = FALSE, colors = NULL, legendTxtSize = 10){

  if(is.null(gene)){
    stop('Please provide a gene name.')
  }

  geneID = gene
  #Protein domain source.
  gff = system.file('extdata', 'protein_domains.txt.gz', package = 'maftools')

  if(Sys.info()[['sysname']] == 'Windows'){
    gff.gz = gzfile(description = gff, open = 'r')
    gff <- suppressWarnings( data.table(read.csv( file = gff.gz, header = TRUE, sep = '\t', stringsAsFactors = FALSE)) )
    close(gff.gz)
  } else{
    gff = data.table::fread(input = paste('zcat <', gff), sep = '\t', stringsAsFactors = FALSE)
  }



  mut = maf@data
  #Remove CNV's
  mut = mut[!Variant_Type %in% 'CNV']

  if(is.null(AACol)){
    if(! 'AAChange' %in% colnames(mut)){
      message('Available fields:')
      print(colnames(mut))
      stop('AAChange field not found in MAF. Use argument AACol to manually specifiy field name containing protein changes.')
    }
  }else{
    colnames(mut)[which(colnames(mut) == AACol)] = 'AAChange'
  }

  prot.dat = mut[Hugo_Symbol %in% geneID, .(Variant_Type, Variant_Classification, AAChange)]
  if(nrow(prot.dat) == 0){
    stop(paste(geneID, 'does not seem to have any mutations!', sep=' '))
  }


  prot = gff[HGNC %in% geneID]

  if(nrow(prot) == 0){
    stop(paste('Structure for protein', geneID, 'not found.', sep=' '))
  }

  if(!is.null(refSeqID)){
    prot = gff[refseq.ID == refSeqID]
  } else if(!is.null(proteinID)){
    prot = gff[protein.ID == proteinID]
  } else{
    txs = unique(prot$refseq.ID)
    if(length(txs) > 1){
      message(paste(length(txs), ' transcripts available. Use arguments refSeqID or proteinID to manually specify tx name.', sep = ''))
      print(prot[,.(HGNC, refseq.ID, protein.ID, aa.length, Start, End, Label)])
      prot = prot[which(prot$aa.length == max(prot$aa.length)),]
      if(length(unique(prot$refseq.ID)) > 1){
        prot = prot[which(prot$refseq.ID == unique(prot[,refseq.ID])[1]),]
        message(paste('Using longer transcript', unique(prot[,refseq.ID])[1], 'for now.', sep=' '))
      } else{
        message(paste('Using longer transcript', unique(prot[,refseq.ID])[1], 'for now.', sep=' '))
      }
    }
  }

  #Legth of protein
  len = as.numeric(max(prot$aa.length))
  #Remove NA's
  prot = prot[!is.na(Label)]

  #hard coded colors for variant classification if user doesnt provide any
  if(is.null(colors)){
    col = c(RColorBrewer::brewer.pal(12,name = "Paired"), RColorBrewer::brewer.pal(11,name = "Spectral")[1:3],'black')
    names(col) = names = c('Nonstop_Mutation','Frame_Shift_Del','Silent','Missense_Mutation','IGR','Nonsense_Mutation',
                           'RNA','Splice_Site','Intron','Frame_Shift_Ins','In_Frame_Dell','In_Frame_Del','ITD','In_Frame_Ins','Translation_Start_Site',"Multi_Hit")
  }else{
    col = colors
  }



  prot.dat = prot.dat[Variant_Classification != 'Splice_Site']
  #Remove 'p.'
  prot.spl = strsplit(x = as.character(prot.dat$AAChange), split = '.', fixed = TRUE)
  prot.conv = sapply(sapply(prot.spl, function(x) x[length(x)]), '[', 1)

  prot.dat[,conv := prot.conv]
  #If conversions are in HGVSp_long (default HGVSp) format, we will remove strings Ter followed by anything (e.g; p.Asn1986GlnfsTer13)
  pos = gsub(pattern = 'Ter.*', replacement = '',x = prot.dat$conv)

  #Following parsing takes care of most of HGVSp_short and HGVSp_long format
  pos = gsub(pattern = '[[:alpha:]]', replacement = '', x = pos)
  pos = gsub(pattern = '\\*$', replacement = '', x = pos) #Remove * if nonsense mutation ends with *
  pos = gsub(pattern = '^\\*', replacement = '', x = pos) #Remove * if nonsense mutation starts with *
  pos = gsub(pattern = '\\*.*', replacement = '', x = pos) #Remove * followed by position e.g, p.C229Lfs*18


  pos = as.numeric(sapply(strsplit(x = pos, split = '_', fixed = TRUE), '[[', 1))
  prot.dat[,pos := pos]

  if(nrow( prot.dat[is.na(pos)]) > 0){
    message(paste('Removed', nrow( prot.dat[is.na(prot.dat$pos),]), 'mutations for which AA position was not available', sep = ' '))
    print(prot.dat[is.na(pos)])
    prot.dat = prot.dat[!is.na(pos)]
  }

  #prot.snp.sumamry = ddply(prot.dat, .variables = c('Variant_Classification', 'conv', 'pos'), summarise, count =length(AAChange))
  prot.snp.sumamry = prot.dat[,.N, .(Variant_Classification, conv, pos)]
  colnames(prot.snp.sumamry)[ncol(prot.snp.sumamry)] = 'count'
  maxCount = max(prot.snp.sumamry$count, na.rm = TRUE)

  prot.snp.sumamry = prot.snp.sumamry[order(prot.snp.sumamry$pos),]
  #prot.snp.sumamry$distance = c(0,diff(prot.snp.sumamry$pos))

  #prot.snp.sumamry$pos2 = ifelse(test = prot.snp.sumamry$distance <= 5 & prot.snp.sumamry$distance != 0, yes = prot.snp.sumamry$pos+5, no = prot.snp.sumamry$pos)
  clusterSize = 10 #Change this later as an argument to user.
  if(repel){
    prot.snp.sumamry = repelPoints(dat = prot.snp.sumamry, protLen = len, clustSize = clusterSize)
  }else{
    prot.snp.sumamry$pos2 = prot.snp.sumamry$pos
  }


  if(max(prot.snp.sumamry$count) > 10){

    #prot.snp.sumamry1 = dplyr::filter(prot.snp.sumamry, count <= 10)
    prot.snp.sumamry1 = prot.snp.sumamry[count <= 10]
    #prot.snp.sumamry2 = dplyr::filter(prot.snp.sumamry, count > 10)
    prot.snp.sumamry2 = prot.snp.sumamry[count > 10]

    if(nrow(prot.snp.sumamry1) == 0){
      maxCount = 5
    }else{
      maxCount = max(prot.snp.sumamry1$count) + 2
    }

    prot.snp.sumamry$count2 = ifelse(test = prot.snp.sumamry$count > 10, no = prot.snp.sumamry$count, yes = maxCount+2)

    if(nrow(prot.snp.sumamry1) > 0){

      p = ggplot()+geom_point(data = prot.snp.sumamry2, aes(x = pos2, y = maxCount+2, color = Variant_Classification), size = 9, alpha = 0.6)+scale_color_manual(values = col)+cowplot::theme_cowplot()+
        theme(legend.text=element_text(size = legendTxtSize), legend.position = 'none', axis.line.x = element_blank(), legend.title = element_blank())+
        scale_y_continuous(breaks = c(0:maxCount, maxCount+3), labels = c(0:maxCount, max(prot.snp.sumamry2$count)+3), limits = c(0, maxCount+3))+xlab('')+ylab('Number of variants')+scale_x_continuous(limits = c(0, len))+
        geom_text(data = prot.snp.sumamry2, aes(x = pos2, y = maxCount+2,label = count))+theme(legend.position = 'none')

      p = p+geom_point(data = prot.snp.sumamry1, aes(x = pos2, y = count, color = Variant_Classification), size = 3, alpha = 0.6)+
        theme(legend.position = 'bottom')+guides(colour = guide_legend(override.aes = list(size=3)))
    }else{
      p = ggplot()+geom_point(data = prot.snp.sumamry2, aes(x = pos2, y = maxCount+2, color = Variant_Classification), size = 9, alpha = 0.6)+scale_color_manual(values = col)+cowplot::theme_cowplot()+
        theme(legend.text=element_text(size = legendTxtSize), legend.position = 'none', axis.line.x = element_blank(), legend.title = element_blank())+
        scale_y_continuous(breaks = c(0:maxCount, maxCount+3), labels = c(0:maxCount, max(prot.snp.sumamry2$count)+3), limits = c(0, maxCount+3))+xlab('')+ylab('Number of variants')+scale_x_continuous(limits = c(0, len))+
        geom_text(data = prot.snp.sumamry2, aes(x = pos2, y = maxCount+2,label = count))+theme(legend.position = 'none')+
        theme(legend.position = 'bottom')+guides(colour = guide_legend(override.aes = list(size=3)))
    }

  } else{

    maxCount = maxCount+1
    prot.snp.sumamry$count2 = prot.snp.sumamry$count
    #Plot points
    p = ggplot()+geom_point(data = prot.snp.sumamry, aes(x = pos2, y = count, color = Variant_Classification), size = 3, alpha = 0.6)+scale_color_manual(values = col)+cowplot::theme_cowplot()+
      theme(legend.position = 'bottom', axis.line.x = element_blank(), legend.title = element_blank())+
      scale_y_continuous(breaks = c(0:maxCount), labels = c(0:maxCount), limits = c(0, maxCount))+xlab('')+ylab('Number of Variants')

  }

  #Plot background protein domain
  p = p+geom_rect(data = prot, aes(xmin = 0, xmax = len, ymin = 0.1, ymax = 0.3), fill = 'gray')

  #Plot protein domains. If no domains found, just draw background protein.
  if(nrow(prot) > 0){
    p = p+geom_rect(data = prot, aes(xmin = Start, xmax = End, ymin = 0.0, ymax = 0.4, fill = Label))
  }

  #Plot lines from protein to points (make lollypops)
  p = p+geom_segment(data = prot.snp.sumamry, aes(x = pos, xend = pos2, y = 0.3, yend = count2-0.03))+ggtitle(label = paste(geneID, ' (',unique(prot[,refseq.ID]), ')',sep=''))

  #If user asks to label points, use ggrepel to label.
  if(!is.null(labelPos)){
    prot.snp.sumamry = data.table(prot.snp.sumamry)

    if(length(labelPos) == 1){

      if(labelPos == 'all'){
        p = p+geom_text_repel(data = prot.snp.sumamry, aes(pos2, count2, label = as.character(conv)), force = 2, nudge_y = 0.6, nudge_x = 0.3)
      }else{
        prot.snp.sumamry$labThis = ifelse(test = prot.snp.sumamry$pos %in% labelPos, yes = 'yes', no = 'no')
        p = p+ggrepel::geom_text_repel(data = prot.snp.sumamry[labThis %in% 'yes'], aes(pos2, count2, label = as.character(conv)), force = 2, nudge_y = 0.6, nudge_x = 0.3)
        #p = p+geom_text_repel(data = prot.snp.summary[labThis %in% 'yes'], aes(pos2, count2, label = as.character(conv)), force = 2, nudge_y = 0.6, nudge_x = 0.3)
      }
    } else{
      prot.snp.sumamry$labThis = ifelse(test = prot.snp.sumamry$pos %in% labelPos, yes = 'yes', no = 'no')
      p = p+ggrepel::geom_text_repel(data = prot.snp.sumamry[labThis %in% 'yes'], aes(pos2, count2, label = as.character(conv)), force = 2, nudge_y = 0.6, nudge_x = 0.3)
      #p = p+geom_text_repel(data = prot.snp.summary[labThis %in% 'yes'], aes(pos2, count2, label = as.character(conv)), force = 2, nudge_y = 0.6, nudge_x = 0.3)
    }
  }
  print(p)
  return(p)
  #return(prot.snp.sumamry)
}
