#' Rainfall plot to display kataegis or hyper mutated genomic regions.
#' @description Plots inter variant distance as a function of genomic locus.
#' @details Note that detected change points are only loci where the distribution of inter-event distance changes. Segments may have to be manually inferred by adjacent change-points.
#'
#' @param maf an \code{\link{MAF}} object generated by \code{\link{read.maf}}. Required.
#' @param tsb specify sample names (Tumor_Sample_Barcodes) for which plotting has to be done. If NULL, draws plot for most mutated sample.
#' @param detectChangePoints If TRUE, detectes genomic change points where potential kataegis are formed. Results are written to an output tab delimted file.
#' @param color named vector of colors for each coversion class.
#' @param ref.build Reference build for chromosome sizes. Can be hg18, hg19 or hg38. Default hg19.
#' @param savePlot If TRUE plot is saved to output pdf. Default FALSE.
#' @param width width of plot to be saved.
#' @param height height of plot to be saved.
#' @param fontSize Default 12.
#' @param pointSize Default 2.
#' @return returns ggplot object of the plot which can be further modified.
#' @importFrom changepoint cpt.mean cpts
#' @export


rainfallPlot = function(maf, tsb = NULL, detectChangePoints = FALSE, ref.build = 'hg19', color = NULL, savePlot = FALSE, width = 6, height = 3, fontSize = 12, pointSize = 1){

  if(is.null(tsb)){
    tsb = as.character(getSampleSummary(maf)[1,Tumor_Sample_Barcode])
  }

  maf.dat = subsetMaf(maf = maf, includeSyn = TRUE, tsb = tsb, fields = 'Hugo_Symbol')
  if(nrow(maf.dat) < 1){
    stop(paste(tsb, 'not found.', sep = ' '))
  }

  maf = maf.dat[,.(Chromosome, Hugo_Symbol, Start_Position, End_Position, Reference_Allele, Tumor_Seq_Allele2, Tumor_Sample_Barcode, Variant_Type)]

  maf.snp = maf[Variant_Type == 'SNP']
  if(nrow(maf) == 0){
    stop('No more single nucleotide variants left after filtering for SNP in Variant_Type field.')
  }

  maf.snp$con = paste(maf.snp[,Reference_Allele], maf.snp[,Tumor_Seq_Allele2], sep = '>')

  maf.snp$con.class = suppressWarnings(as.character(factor(maf.snp$con, levels = c("A>G", "T>C", "C>T", "G>A", "A>T", "T>A", "A>C", "T>G", "C>A", "G>T", "C>G", "G>C"),
                                                                   labels = c("T>C", "T>C", "C>T", "C>T", "T>A", "T>A", "T>G", "T>G", "C>A", "C>A", "C>G", "C>G"))))

  if(nrow(maf.snp) == 0){
    stop(paste0('No SNVs found in sample ', tsb))
  }

  maf.snp = transformSegments(maf.snp, build = ref.build)
  maf.snp$diff = suppressWarnings( log10(c(0, diff(maf.snp$Start_Position_updated))+1) )
  #Remove any NA's if generated
  maf.snp = maf.snp[complete.cases(diff)]

  if(is.null(color)){
    col = RColorBrewer::brewer.pal(n = 6, name = 'Set1')
    names(col) = c('C>T', 'C>G', 'C>A', 'T>C', 'T>A', 'T>G')
  }else{
    col = color
  }

  #hg19 chromosome lengths
  chr.lens = c(249250621, 243199373, 198022430, 191154276, 180915260, 171115067, 159138663,
               146364022, 141213431, 135534747, 135006516, 133851895, 115169878, 107349540,
               102531392, 90354753, 81195210, 78077248, 59128983, 63025520, 48129895, 51304566,
               155270560, 59373566)

  chr.lens.sumsum = cumsum(chr.lens)

  gg.rf = ggplot(data = maf.snp, aes(x= Start_Position_updated, y = diff, col = con.class))+
    geom_point(size = pointSize, alpha = 0.6)+cowplot::theme_cowplot(font_size = fontSize)+cowplot::background_grid(major = 'y')+xlab('')+ylab('log10(inter event distance)')+
    theme(axis.line.x = element_blank())+scale_x_continuous(breaks = chr.lens.sumsum, labels = c(1:22, 'X', 'Y'))+
    geom_vline(xintercept = chr.lens.sumsum, linetype = 'dotted', size = 0.3)+theme(legend.position = 'bottom', legend.title = element_blank())+
    scale_color_manual(values = col)+ggtitle(tsb)+guides(colour = guide_legend(override.aes = list(size=3)))


  if(detectChangePoints){
    maf.cpt = changepoint::cpt.mean(data = maf.snp$diff, method = 'PELT', class = TRUE, minseglen = 10)
    if(length(cpts(maf.cpt)) == 0){
      message('No changepoints detected!')
    }else{
      maf.snp = maf.snp[!diff %in% 0]
      maf.snp[,min := min(diff), by = .(Chromosome)]
      maf.cpt = maf.snp[changepoint::cpts(object = maf.cpt)]

      gg.rf = gg.rf+geom_segment(aes(x = Start_Position_updated, y = 0.1, yend = min-0.2, xend = End_Position_updated), data = maf.cpt, inherit.aes = FALSE, arrow = arrow(length = unit(0.2, "cm")))
      maf.cpt = maf.cpt[,.(Chromosome, Start_Position)]
      colnames(maf.cpt) = c('Chromosome', 'Position')
      message('Change points detected at:')
      print(maf.cpt)
      write.table(x = maf.cpt, file = paste(tsb, 'changePoints.tsv', sep='_'), sep='\t', quote = FALSE, row.names = FALSE)
    }
  }


  if(savePlot){
#     gg.rf = ggplot(data = maf.snp, aes(x= Start_Position_updated, y = diff, col = con.class))+
#       geom_point(size = pointSize, alpha = 0.6)+cowplot::theme_cowplot(font_size = fontSize)+cowplot::background_grid(major = 'y')+xlab('')+ylab('log10(inter event distance)')+
#       theme(axis.line.x = element_blank())+scale_x_continuous(breaks = chr.lens.sumsum, labels = c(1:22, 'X', 'Y'))+
#       geom_vline(xintercept = chr.lens.sumsum, linetype = 'dotted', size = 0.3)+theme(legend.position = 'bottom', legend.title = element_blank())+
#       scale_color_manual(values = col)+ggtitle(tsb)

    cowplot::save_plot(filename = paste(tsb, 'rainfallPlot.pdf', sep = '_'), plot = gg.rf, base_height = height, base_width = width)
  }

  print(gg.rf)
  return(gg.rf)
}
